<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>tacco.tools._orthology &mdash; TACCO  documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../../../_static/sg_gallery.css?v=ad0ac74c" />
      <link rel="stylesheet" type="text/css" href="../../../_static/readthedocs-custom.css?v=c8300288" />
      <link rel="stylesheet" type="text/css" href="../../../_static/notebook_hacks.css?v=e4ce2130" />

  
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../../../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../../../_static/doctools.js?v=888ff710"></script>
        <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html">
            
              <img src="../../../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../index.html">Home</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../_autosummary/tacco.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">TACCO</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">tacco.tools._orthology</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for tacco.tools._orthology</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="kn">from</span> <span class="nn">._goa</span> <span class="kn">import</span> <span class="n">_get_downloaded_file</span>
<span class="kn">from</span> <span class="nn">..utils</span> <span class="kn">import</span> <span class="n">row_scale</span><span class="p">,</span> <span class="n">col_scale</span><span class="p">,</span> <span class="n">get_sum</span>

<span class="n">_homology_file</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">_gene2DB_matrices</span> <span class="o">=</span> <span class="kc">None</span>
<span class="n">_synonym2gene_matrices</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="setup_orthology_converter">
<a class="viewcode-back" href="../../../_autosummary/tacco.tools.setup_orthology_converter.html#tacco.tools.setup_orthology_converter">[docs]</a>
<span class="k">def</span> <span class="nf">setup_orthology_converter</span><span class="p">(</span>
    <span class="n">homology_file</span><span class="o">=</span><span class="s1">&#39;http://www.informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt&#39;</span><span class="p">,</span>
    <span class="n">working_directory</span><span class="o">=</span><span class="s1">&#39;.&#39;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;\</span>
<span class="sd">    Setup for converting gene names between species using MGI homology</span>
<span class="sd">    information (Mouse Genome Informatics Web Site,</span>
<span class="sd">    `http://www.informatics.jax.org/homology.shtml`).</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    homology_file</span>
<span class="sd">        File containing a mapping from species specific gene symbols to common</span>
<span class="sd">        MGI &quot;DB class key&quot;/&quot;HomoloGene ID&quot;, e.g. downloaded from `http://www.informatics.jax.org/downloads/reports/HOM_AllOrganism.rpt`.</span>
<span class="sd">        If this is not available as a local file, it is treated as an URL and</span>
<span class="sd">        downloaded to the `working_directory` if necessary, see below.</span>
<span class="sd">    working_directory</span>
<span class="sd">        Directory where to buffer downloaded files. If a file of the same name</span>
<span class="sd">        already exists in this directory, it is not downloaded again.</span>
<span class="sd">        </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    A string containing the local path to the homology file.</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">global</span> <span class="n">_homology_file</span>
    
    <span class="c1"># download files if necessary</span>
    <span class="n">_homology_file</span> <span class="o">=</span> <span class="n">_get_downloaded_file</span><span class="p">(</span><span class="n">homology_file</span><span class="p">,</span> <span class="n">working_directory</span><span class="p">)</span>
    
    <span class="k">return</span> <span class="n">_homology_file</span></div>


<span class="k">def</span> <span class="nf">_get_homology_key</span><span class="p">(</span><span class="n">hom_df</span><span class="p">):</span>
    <span class="n">homology_key</span> <span class="o">=</span> <span class="s1">&#39;DB Class Key&#39;</span>
    <span class="k">if</span> <span class="n">homology_key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hom_df</span><span class="p">:</span>
        <span class="n">homology_key</span> <span class="o">=</span> <span class="s1">&#39;HomoloGene ID&#39;</span>
    <span class="k">if</span> <span class="n">homology_key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hom_df</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Neither &quot;DB Class Key&quot; nor &quot;HomoloGene ID&quot; are available in the file &quot;</span><span class="si">{</span><span class="n">_homology_file</span><span class="si">}</span><span class="s1">&quot;!&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">homology_key</span>

<span class="k">def</span> <span class="nf">_construct_gene2DB_matrix</span><span class="p">(</span><span class="n">tax_id</span><span class="p">):</span>
    
    <span class="k">global</span> <span class="n">_homology_file</span>
    <span class="k">if</span> <span class="n">_homology_file</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;&quot;run_orthology_converter&quot; was run before &quot;setup_orthology_converter&quot;! Running &quot;setup_orthology_converter&quot; now with default arguments...&#39;</span><span class="p">)</span>
        <span class="n">setup_orthology_converter</span><span class="p">()</span>
    
    <span class="n">full_hom_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">_homology_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    
    <span class="n">homology_key</span> <span class="o">=</span> <span class="n">_get_homology_key</span><span class="p">(</span><span class="n">full_hom_df</span><span class="p">)</span>
                         
    <span class="n">full_hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">full_hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span> <span class="c1"># consistent categories across all species</span>
    
    <span class="n">sub_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">,</span><span class="n">homology_key</span><span class="p">,]</span>
    <span class="k">if</span> <span class="s1">&#39;Synonyms&#39;</span> <span class="ow">in</span> <span class="n">full_hom_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">sub_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">)</span>
    
    <span class="n">hom_df</span> <span class="o">=</span> <span class="n">full_hom_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">full_hom_df</span><span class="p">[</span><span class="s1">&#39;NCBI Taxon ID&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">tax_id</span><span class="p">,</span> <span class="n">sub_cols</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">hom_df</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The tax_id </span><span class="si">{</span><span class="n">tax_id</span><span class="si">!r}</span><span class="s1"> is not available in the file </span><span class="si">{</span><span class="n">_homology_file</span><span class="si">!r}</span><span class="s1">!&#39;</span><span class="p">)</span>
    
    <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span> <span class="c1"># consistent categories only within the selected species</span>
    
    <span class="c1"># create an anndata containing the mapping from gene symbol synonyms to gene symbols</span>
    <span class="k">if</span> <span class="s1">&#39;Synonyms&#39;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">hom_df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span>
    
    <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">where</span><span class="p">(</span><span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">(),</span> <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">],</span> <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;|&#39;</span> <span class="o">+</span> <span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">))</span>
    <span class="n">synonym_map</span> <span class="o">=</span> <span class="n">hom_df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;Symbol&#39;</span><span class="p">)[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">explode</span><span class="p">()</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
    <span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span> <span class="c1"># consistent categories only within the selected species</span>
    
    <span class="n">synonym_matrix</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">coo_matrix</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">synonym_map</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">),(</span><span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">codes</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span><span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">codes</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())),</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)))</span>
    <span class="n">synonym_matrix</span> <span class="o">=</span> <span class="n">synonym_matrix</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()</span>
    
    <span class="n">synonym_adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">synonym_matrix</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Synonyms&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">),</span> <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">synonym_map</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">))</span>
    
    <span class="c1"># create an anndata containing the mapping from gene symbols to DB Class Keys/HomoloGene ID</span>
    <span class="n">hom_df</span> <span class="o">=</span> <span class="n">hom_df</span><span class="p">[</span><span class="n">hom_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[:</span><span class="mi">2</span><span class="p">]]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="n">merge_matrix</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">coo_matrix</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hom_df</span><span class="p">),</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">),(</span><span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">codes</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">(),</span><span class="n">hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">codes</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">())),</span><span class="n">shape</span><span class="o">=</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">),</span><span class="nb">len</span><span class="p">(</span><span class="n">hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">)))</span>
    <span class="n">merge_matrix</span> <span class="o">=</span> <span class="n">merge_matrix</span><span class="o">.</span><span class="n">tocsr</span><span class="p">()</span>
    
    <span class="n">merge_adata</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">merge_matrix</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">hom_df</span><span class="p">[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">),</span> <span class="n">var</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span><span class="p">))</span>
    
    <span class="n">hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">hom_df</span><span class="p">[</span><span class="n">homology_key</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">merge_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">tax_id</span><span class="si">}</span><span class="s1"> orthologs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hom_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="n">homology_key</span><span class="p">)[</span><span class="s1">&#39;Symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
    <span class="k">for</span> <span class="n">dbck</span> <span class="ow">in</span> <span class="n">merge_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">merge_adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">tax_id</span><span class="si">}</span><span class="s1"> orthologs&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span><span class="o">.</span><span class="n">index</span><span class="p">:</span>
        <span class="n">merge_adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">dbck</span><span class="p">,</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">tax_id</span><span class="si">}</span><span class="s1"> orthologs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
    
    <span class="k">return</span> <span class="p">(</span> <span class="n">merge_adata</span><span class="p">,</span> <span class="n">synonym_adata</span> <span class="p">)</span>

<span class="k">def</span> <span class="nf">_get_gene2DB_matrix</span><span class="p">(</span><span class="n">tax_id</span><span class="p">):</span>
    <span class="k">global</span> <span class="n">_gene2DB_matrices</span>
    <span class="k">if</span> <span class="n">_gene2DB_matrices</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">_gene2DB_matrices</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">if</span> <span class="n">tax_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">_gene2DB_matrices</span><span class="p">:</span>
        <span class="n">_gene2DB_matrices</span><span class="p">[</span><span class="n">tax_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">_construct_gene2DB_matrix</span><span class="p">(</span><span class="n">tax_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">_gene2DB_matrices</span><span class="p">[</span><span class="n">tax_id</span><span class="p">]</span>

<span class="n">_name_to_tax_id_map</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;human&#39;</span><span class="p">:</span> <span class="mi">9606</span><span class="p">,</span>
    <span class="s1">&#39;mouse&#39;</span><span class="p">:</span> <span class="mi">10090</span><span class="p">,</span>
<span class="p">}</span>
<span class="k">def</span> <span class="nf">_convert_identifier_to_tax_id</span><span class="p">(</span><span class="n">identifier</span><span class="p">):</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">identifier</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">identifier</span> <span class="ow">in</span> <span class="n">_name_to_tax_id_map</span><span class="p">:</span>
            <span class="n">identifier</span> <span class="o">=</span> <span class="n">_name_to_tax_id_map</span><span class="p">[</span><span class="n">identifier</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;The identifier </span><span class="si">{</span><span class="n">identifier</span><span class="si">!r}</span><span class="s1"> is not supported as an alias of a taxonomy id. The currently supported aliases are:</span><span class="se">\n</span><span class="si">{</span><span class="n">_name_to_tax_id_map</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">identifier</span>

<div class="viewcode-block" id="run_orthology_converter">
<a class="viewcode-back" href="../../../_autosummary/tacco.tools.run_orthology_converter.html#tacco.tools.run_orthology_converter">[docs]</a>
<span class="k">def</span> <span class="nf">run_orthology_converter</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">source_tax_id</span><span class="p">,</span>
    <span class="n">target_tax_id</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">target_gene_symbols</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">use_synonyms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;\</span>
<span class="sd">    Run orthology conversion between species using MGI homology information</span>
<span class="sd">    information (Mouse Genome Informatics Web Site,</span>
<span class="sd">    `http://www.informatics.jax.org/homology.shtml`). The function works on the</span>
<span class="sd">    var.index of the adata and assumes to find gene symbols there.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    adata</span>
<span class="sd">        An :class:`~anndata.AnnData` containing the data to convert. If `None`,</span>
<span class="sd">        return an :class:`~anndata.AnnData` containing the conversion matrix.</span>
<span class="sd">    source_tax_id</span>
<span class="sd">        The NCBI taxonomy ID of the source data. Some strings can be used as</span>
<span class="sd">        aliases of their taxonomy IDs (e.g. &#39;human&#39;=9606 and &#39;mouse&#39;=10090).</span>
<span class="sd">    target_tax_id</span>
<span class="sd">        The target NCBI taxonomy ID. Some strings can be used as aliases of</span>
<span class="sd">        their taxonomy IDs (e.g. &#39;human&#39;=9606 and &#39;mouse&#39;=10090). If `None`,</span>
<span class="sd">        convert to the common MGI &quot;DB class key&quot;/&quot;HomoloGene ID&quot; (recommended, as this avoids</span>
<span class="sd">        extra ambuigities in multi-to-multi mappings).</span>
<span class="sd">    target_gene_symbols</span>
<span class="sd">        The gene symbols appearing in the target data. Supplying this can make</span>
<span class="sd">        a difference if the symbols appearing there are contained in the</span>
<span class="sd">        &quot;Synonyms&quot; column of the homology information. Synonyms are</span>
<span class="sd">        unambiguously taken care of on the source side.</span>
<span class="sd">    use_synonyms</span>
<span class="sd">        Whether to use the &quot;Synonyms&quot; column from the homology information.</span>
<span class="sd">        </span>
<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    Returns a :class:`~anndata.AnnData` containing the result of the\</span>
<span class="sd">    orthology conversion.</span>
<span class="sd">    </span>
<span class="sd">    &quot;&quot;&quot;</span>
    
    <span class="n">source_tax_id</span> <span class="o">=</span> <span class="n">_convert_identifier_to_tax_id</span><span class="p">(</span><span class="n">source_tax_id</span><span class="p">)</span>
    
    <span class="n">merge_matrix</span><span class="p">,</span> <span class="n">synonym_matrix</span> <span class="o">=</span> <span class="n">_get_gene2DB_matrix</span><span class="p">(</span><span class="n">source_tax_id</span><span class="p">)</span>
    
    <span class="c1"># normalize the merge matrix to account for many-to-many mappings</span>
    <span class="n">merge_matrix</span> <span class="o">=</span> <span class="n">merge_matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># dont change the original</span>
    <span class="n">row_scale</span><span class="p">(</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="n">get_sum</span><span class="p">(</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
    
    <span class="k">if</span> <span class="n">use_synonyms</span><span class="p">:</span>
        <span class="c1"># normalize the synonym matrix separately as the interpretation differs: uncertain many-to-many mapping of different entities VS many-to-many mapping of identical entities...</span>
        <span class="n">synonym_matrix</span> <span class="o">=</span> <span class="n">synonym_matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># dont change the original</span>
        <span class="n">row_scale</span><span class="p">(</span><span class="n">synonym_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="n">get_sum</span><span class="p">(</span><span class="n">synonym_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">))</span>
    
        <span class="c1"># combine the synonym and the merge matrices on the source side</span>
        <span class="n">merge_matrix</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">synonym_matrix</span><span class="o">.</span><span class="n">X</span> <span class="o">@</span> <span class="n">merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">synonym_matrix</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">target_tax_id</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">target_tax_id</span> <span class="o">=</span> <span class="n">_convert_identifier_to_tax_id</span><span class="p">(</span><span class="n">target_tax_id</span><span class="p">)</span>
        <span class="n">target_merge_matrix</span><span class="p">,</span> <span class="n">target_synonym_matrix</span> <span class="o">=</span> <span class="n">_get_gene2DB_matrix</span><span class="p">(</span><span class="n">target_tax_id</span><span class="p">)</span>
        <span class="n">target_merge_matrix</span> <span class="o">=</span> <span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># dont change the original</span>
        <span class="c1"># transposed normalization on the target side, also separate: account for the distribution of counts to genes which are just not appearing in the measurement</span>
        <span class="n">col_scale</span><span class="p">(</span><span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="n">get_sum</span><span class="p">(</span><span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
        
        <span class="k">if</span> <span class="n">use_synonyms</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">target_gene_symbols</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># subset to genes actually appearing</span>
                <span class="n">mapped_genes</span> <span class="o">=</span> <span class="n">target_synonym_matrix</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">target_gene_symbols</span><span class="p">)</span>
                <span class="n">target_synonym_matrix</span> <span class="o">=</span> <span class="n">target_synonym_matrix</span><span class="p">[</span><span class="n">mapped_genes</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># dont change the original</span>
            <span class="c1"># normalize the synonym matrix after potential subsetting to the synonyms which actually appear in the dataset</span>
            <span class="n">col_scale</span><span class="p">(</span><span class="n">target_synonym_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="mi">1</span><span class="o">/</span><span class="n">get_sum</span><span class="p">(</span><span class="n">target_synonym_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">))</span>
            <span class="n">target_merge_matrix</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">target_synonym_matrix</span><span class="o">.</span><span class="n">X</span> <span class="o">@</span> <span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">target_synonym_matrix</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">var</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">target_gene_symbols</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># subset to genes actually appearing</span>
                <span class="n">mapped_genes</span> <span class="o">=</span> <span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">target_gene_symbols</span><span class="p">)</span>
                <span class="n">target_merge_matrix</span> <span class="o">=</span> <span class="n">target_merge_matrix</span><span class="p">[</span><span class="n">mapped_genes</span><span class="p">]</span>
            
        <span class="n">merge_matrix</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">X</span> <span class="o">@</span> <span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">target_merge_matrix</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
    
    <span class="k">if</span> <span class="n">adata</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">merge_matrix</span>
    
    <span class="c1"># subset to genes actually appearing</span>
    <span class="n">mapped_genes</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
    <span class="n">adata</span><span class="p">,</span> <span class="n">merge_matrix</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[:,</span><span class="n">mapped_genes</span><span class="p">],</span> <span class="n">merge_matrix</span><span class="p">[</span><span class="n">mapped_genes</span><span class="p">]</span>
    
    <span class="k">return</span> <span class="n">ad</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span> <span class="o">@</span> <span class="n">merge_matrix</span><span class="o">.</span><span class="n">X</span><span class="p">,</span> <span class="n">obs</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">obsm</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">merge_matrix</span><span class="o">.</span><span class="n">var</span><span class="p">)</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>