<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>In-silico mixed Mouse Colon scRNA-seq data &mdash; TACCO  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=ad0ac74c" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
      <link rel="stylesheet" type="text/css" href="../_static/readthedocs-custom.css?v=c8300288" />
      <link rel="stylesheet" type="text/css" href="../_static/notebook_hacks.css?v=e4ce2130" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Mapping single cells into space" href="mapping_single_cells_into_space.html" />
    <link rel="prev" title="Co-occurrence computations" href="cooccurrence.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="benchmarking.html">Benchmarking runtime and memory of annotation methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="cooccurrence.html">Co-occurrence computations</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">In-silico mixed Mouse Colon scRNA-seq data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Load-data">Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Mix-data-in-silico">Mix data in-silico</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plotting-options">Plotting options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Visualize-scRNA-seq-data">Visualize scRNA-seq data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Visualize-in-silico-mixtures">Visualize in-silico mixtures</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Benchmark-annotation-methods-on-the-in-silico-mixtures-with-known-ground-truth">Benchmark annotation methods on the in-silico mixtures with known ground truth</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Observation-splitting">Observation splitting</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Annotate-the-spatial-data-with-compositions-of-cell-types">Annotate the spatial data with compositions of cell types</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Split-the-mixed-observations-into-pure-contributions">Split the mixed observations into pure contributions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Visualize-the-split-observations">Visualize the split observations</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="mapping_single_cells_into_space.html">Mapping single cells into space</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_splitting.html">Object splitting for in-silico mixed scRNA-seq data and Slide-Seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="osmFISH_single_molecule_annotation.html">Single molecule data: osmFISH mouse somato-sensory cortex</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html">TACCO overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="parameter_variations.html">Parameter Variations for annotation using Optimal Transport</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulated_ambient.html">Simulated single cell expression data with ambient contamination</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulated_dropout.html">Simulated single cell expression data with dropout</a></li>
<li class="toctree-l2"><a class="reference internal" href="single_cell_differentiation.html">Single-cell differentiation</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_colon.html">Slide-Seq Mouse Colon</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_olfactory_bulb.html">Slide-Seq Mouse Olfactory Bulb - multiple pucks</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_olfactory_bulb_single.html">Slide-Seq Mouse Olfactory Bulb - single puck</a></li>
<li class="toctree-l2"><a class="reference internal" href="visium_mouse_kidney.html">Visium Mouse Kidney</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../_autosummary/tacco.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TACCO</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../examples.html">Examples</a></li>
      <li class="breadcrumb-item active">In-silico mixed Mouse Colon scRNA-seq data</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="In-silico-mixed-Mouse-Colon-scRNA-seq-data">
<h1>In-silico mixed Mouse Colon scRNA-seq data<a class="headerlink" href="#In-silico-mixed-Mouse-Colon-scRNA-seq-data" title="Link to this heading">Â¶</a></h1>
<p>This example uses TACCO to annotate and analyse in-silico mixtures of mouse colon scRNA-seq data (Avraham-Davidi et al.).</p>
<p>(Avraham-Davidi et al.): Avraham-Davidi I, Mages S, Klughammer J, et al. Integrative single cell and spatial transcriptomics of colorectal cancer reveals multicellular functional units that support tumor progression. doi: <a class="reference external" href="https://doi.org/10.1101/2022.10.02.508492">https://doi.org/10.1101/2022.10.02.508492</a></p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>

<span class="kn">import</span> <span class="nn">tacco</span> <span class="k">as</span> <span class="nn">tc</span>

<span class="c1"># The notebook expects to be executed either in the workflow directory or in the repository root folder...</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;workflow&#39;</span> <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;workflow/common_code.py&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;..&#39;</span><span class="p">))</span>
<span class="kn">import</span> <span class="nn">common_code</span>
</pre></div>
</div>
</div>
<section id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Link to this heading">Â¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reference_data_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/slideseq_mouse_colon/data&#39;</span><span class="p">)</span>
<span class="n">plot_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/insilico_mouse_colon&#39;</span><span class="p">,</span> <span class="n">create_if_not_existent</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">env_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/env_links&#39;</span><span class="p">)</span>
<span class="n">reference</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">reference_data_path</span><span class="si">}</span><span class="s1">/scrnaseq.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Mix-data-in-silico">
<h2>Mix data in-silico<a class="headerlink" href="#Mix-data-in-silico" title="Link to this heading">Â¶</a></h2>
<p>Generate in-silico mixtures of scRNA-seq data to benchmark methods with a known ground truth</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">bead_sizes</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.5</span><span class="p">,</span><span class="mf">1.0</span><span class="p">,</span><span class="mf">1.5</span><span class="p">,</span><span class="mf">2.0</span><span class="p">]</span>

<span class="n">capture_rate</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="n">bead_shape</span> <span class="o">=</span> <span class="s1">&#39;gauss&#39;</span>
<span class="n">ntdata_max</span> <span class="o">=</span> <span class="mi">10</span><span class="o">**</span><span class="mi">4</span>

<span class="n">ref_annotation_key</span> <span class="o">=</span> <span class="s1">&#39;labels&#39;</span>
<span class="n">tdata_annotation_key</span> <span class="o">=</span> <span class="s1">&#39;reads_&#39;</span> <span class="o">+</span> <span class="n">ref_annotation_key</span>

<span class="n">sdatas</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">bead_size</span> <span class="ow">in</span> <span class="n">bead_sizes</span><span class="p">:</span>
    <span class="n">sdata</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">mix_in_silico</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">type_key</span><span class="o">=</span><span class="n">ref_annotation_key</span><span class="p">,</span> <span class="n">n_samples</span><span class="o">=</span><span class="n">ntdata_max</span><span class="p">,</span> <span class="n">bead_shape</span><span class="o">=</span><span class="n">bead_shape</span><span class="p">,</span> <span class="n">bead_size</span><span class="o">=</span><span class="n">bead_size</span><span class="p">,</span> <span class="n">capture_rate</span><span class="o">=</span><span class="n">capture_rate</span><span class="p">,)</span>
    <span class="n">sdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">tdata_annotation_key</span><span class="p">]</span> <span class="o">/=</span> <span class="n">sdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">tdata_annotation_key</span><span class="p">]</span><span class="o">.</span><span class="n">to_numpy</span><span class="p">()</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)[:,</span><span class="kc">None</span><span class="p">]</span>
    <span class="n">sdatas</span><span class="p">[</span><span class="n">bead_size</span><span class="p">]</span> <span class="o">=</span> <span class="n">sdata</span>
</pre></div>
</div>
</div>
</section>
<section id="Plotting-options">
<h2>Plotting options<a class="headerlink" href="#Plotting-options" title="Link to this heading">Â¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">highres</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">default_dpi</span> <span class="o">=</span> <span class="mf">100.0</span>
<span class="k">if</span> <span class="n">highres</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">648.0</span>
    <span class="n">hr_ext</span> <span class="o">=</span> <span class="s1">&#39;_hd&#39;</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">default_dpi</span>
    <span class="n">hr_ext</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

<span class="n">axsize</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span><span class="o">*</span><span class="mf">0.5</span>

<span class="n">labels_colors</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">({</span><span class="s1">&#39;Epi&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.00784313725490196</span><span class="p">,</span> <span class="mf">0.24313725490196078</span><span class="p">,</span> <span class="mf">1.0</span><span class="p">),</span> <span class="s1">&#39;B&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.10196078431372549</span><span class="p">,</span> <span class="mf">0.788235294117647</span><span class="p">,</span> <span class="mf">0.2196078431372549</span><span class="p">),</span> <span class="s1">&#39;TNK&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">1.0</span><span class="p">,</span> <span class="mf">0.48627450980392156</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">),</span> <span class="s1">&#39;Mono&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.5490196078431373</span><span class="p">,</span> <span class="mf">0.03137254901960784</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">),</span> <span class="s1">&#39;Mac&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.9098039215686274</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">,</span> <span class="mf">0.043137254901960784</span><span class="p">),</span> <span class="s1">&#39;Gran&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.34901960784313724</span><span class="p">,</span> <span class="mf">0.11764705882352941</span><span class="p">,</span> <span class="mf">0.44313725490196076</span><span class="p">),</span> <span class="s1">&#39;Mast&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.23529411764705882</span><span class="p">,</span> <span class="mf">0.23529411764705882</span><span class="p">,</span> <span class="mf">0.23529411764705882</span><span class="p">),</span> <span class="s1">&#39;Endo&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.8549019607843137</span><span class="p">,</span> <span class="mf">0.5450980392156862</span><span class="p">,</span> <span class="mf">0.7647058823529411</span><span class="p">),</span> <span class="s1">&#39;Fibro&#39;</span><span class="p">:</span> <span class="p">(</span><span class="mf">0.6235294117647059</span><span class="p">,</span> <span class="mf">0.2823529411764706</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">)})</span>
</pre></div>
</div>
</div>
</section>
<section id="Visualize-scRNA-seq-data">
<h2>Visualize scRNA-seq data<a class="headerlink" href="#Visualize-scRNA-seq-data" title="Link to this heading">Â¶</a></h2>
<p>Create UMAP for the scRNA-seq data</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ref_umap</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">umap_single_cell_data</span><span class="p">(</span><span class="n">reference</span><span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">ref_umap</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;labels&#39;</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">labels_colors</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">axes_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP 0&#39;</span><span class="p">,</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SCumap...SCprep...time 10.525800943374634
time 80.12539029121399
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_insilico_mouse_colon_12_1.png" src="../_images/notebooks_insilico_mouse_colon_12_1.png" />
</div>
</div>
</section>
<section id="Visualize-in-silico-mixtures">
<h2>Visualize in-silico mixtures<a class="headerlink" href="#Visualize-in-silico-mixtures" title="Link to this heading">Â¶</a></h2>
<p>Create UMAP for the in-silico mixtures</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdata</span> <span class="o">=</span> <span class="n">sdatas</span><span class="p">[</span><span class="mf">1.0</span><span class="p">]</span>
<span class="n">tdata_umap</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">umap_single_cell_data</span><span class="p">(</span><span class="n">tdata</span><span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">tdata_umap</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">tdata_annotation_key</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">labels_colors</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">axes_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP 0&#39;</span><span class="p">,</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SCumap...SCprep...time 6.028910398483276
time 46.6459801197052
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_insilico_mouse_colon_15_1.png" src="../_images/notebooks_insilico_mouse_colon_15_1.png" />
</div>
</div>
</section>
<section id="Benchmark-annotation-methods-on-the-in-silico-mixtures-with-known-ground-truth">
<h2>Benchmark annotation methods on the in-silico mixtures with known ground truth<a class="headerlink" href="#Benchmark-annotation-methods-on-the-in-silico-mixtures-with-known-ground-truth" title="Link to this heading">Â¶</a></h2>
<p>Define parameters for the annotation methods to use</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">methods</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;NMFreg&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;NMFreg&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;RCTD&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;RCTD&#39;</span><span class="p">,</span> <span class="s1">&#39;conda_env&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">env_path</span><span class="si">}</span><span class="s1">/RCTD_env&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;SVM&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span><span class="s1">&#39;svm&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;SingleR&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span><span class="s1">&#39;SingleR&#39;</span><span class="p">,</span> <span class="s1">&#39;conda_env&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">env_path</span><span class="si">}</span><span class="s1">/SingleR_env&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;WOT&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;WOT&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;TACCO&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;OT&#39;</span><span class="p">,</span> <span class="s1">&#39;multi_center&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,},</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">bead_size</span> <span class="ow">in</span> <span class="n">bead_sizes</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span><span class="n">params</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;running method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s1"> for bead size </span><span class="si">{</span><span class="n">bead_size</span><span class="si">}</span><span class="s1"> ...&#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="n">results</span><span class="p">[(</span><span class="n">bead_size</span><span class="p">,</span><span class="n">method</span><span class="p">)]</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">benchmarking</span><span class="o">.</span><span class="n">benchmark_annotate</span><span class="p">(</span><span class="n">sdatas</span><span class="p">[</span><span class="n">bead_size</span><span class="p">],</span><span class="n">reference</span><span class="p">,</span><span class="n">annotation_key</span><span class="o">=</span><span class="s1">&#39;labels&#39;</span><span class="p">,</span><span class="o">**</span><span class="n">params</span><span class="p">);</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;done&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
running method NMFreg for bead size 0.5 ...done
running method RCTD for bead size 0.5 ...done
running method SVM for bead size 0.5 ...done
running method SingleR for bead size 0.5 ...done
running method WOT for bead size 0.5 ...done
running method TACCO for bead size 0.5 ...done
running method NMFreg for bead size 1.0 ...done
running method RCTD for bead size 1.0 ...done
running method SVM for bead size 1.0 ...done
running method SingleR for bead size 1.0 ...done
running method WOT for bead size 1.0 ...done
running method TACCO for bead size 1.0 ...done
running method NMFreg for bead size 1.5 ...done
running method RCTD for bead size 1.5 ...done
running method SVM for bead size 1.5 ...done
running method SingleR for bead size 1.5 ...done
running method WOT for bead size 1.5 ...done
running method TACCO for bead size 1.5 ...done
running method NMFreg for bead size 2.0 ...done
running method RCTD for bead size 2.0 ...done
running method SVM for bead size 2.0 ...done
running method SingleR for bead size 2.0 ...done
running method WOT for bead size 2.0 ...done
running method TACCO for bead size 2.0 ...done
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="p">(</span><span class="n">bead_size</span><span class="p">,</span><span class="n">method</span><span class="p">),</span><span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">unused_key</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">find_unused_key</span><span class="p">(</span><span class="n">sdatas</span><span class="p">[</span><span class="n">bead_size</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">)</span>
    <span class="n">sdatas</span><span class="p">[</span><span class="n">bead_size</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">unused_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[(</span><span class="n">bead_size</span><span class="p">,</span><span class="n">method</span><span class="p">)][</span><span class="s1">&#39;annotation&#39;</span><span class="p">]</span>
    <span class="n">L2</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">ev</span><span class="o">.</span><span class="n">compute_err</span><span class="p">(</span><span class="n">sdatas</span><span class="p">[</span><span class="n">bead_size</span><span class="p">],</span> <span class="n">unused_key</span><span class="p">,</span> <span class="n">tdata_annotation_key</span><span class="p">,</span> <span class="n">err_method</span><span class="o">=</span><span class="s1">&#39;lp&#39;</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="mi">2</span><span class="p">)[</span><span class="n">unused_key</span><span class="p">]</span>
    <span class="k">del</span> <span class="n">sdatas</span><span class="p">[</span><span class="n">bead_size</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">unused_key</span><span class="p">]</span>
    <span class="n">results</span><span class="p">[(</span><span class="n">bead_size</span><span class="p">,</span><span class="n">method</span><span class="p">)][</span><span class="s1">&#39;L2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">L2</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
    <span class="p">[</span><span class="n">bead_size</span><span class="p">,</span><span class="n">method</span><span class="p">,</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;L2&#39;</span><span class="p">],</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;max_mem_usage_GB&#39;</span><span class="p">],</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;benchmark_time_s&#39;</span><span class="p">]]</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">bead_size</span><span class="p">,</span><span class="n">method</span><span class="p">),</span><span class="n">v</span> <span class="ow">in</span> <span class="n">results</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">],</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;beadsize&#39;</span><span class="p">,</span><span class="s1">&#39;method&#39;</span><span class="p">,</span><span class="s1">&#39;L2 error&#39;</span><span class="p">,</span><span class="s1">&#39;memory (GB)&#39;</span><span class="p">,</span><span class="s1">&#39;time (s)&#39;</span><span class="p">])</span>
<span class="n">quantities</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">res_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;beadsize&#39;</span><span class="p">,</span><span class="s1">&#39;method&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="n">methods</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="s1">&#39;method&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">axs</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">quantities</span><span class="p">),</span> <span class="n">axsize</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span><span class="o">*</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">x_padding</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">y_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="p">:</span><span class="n">common_code</span><span class="o">.</span><span class="n">method_color</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">}</span>
<span class="n">styles</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="p">:</span><span class="n">common_code</span><span class="o">.</span><span class="n">method_style</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">}</span>
<span class="n">res_df</span> <span class="o">=</span> <span class="n">res_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">res_df</span><span class="p">[</span><span class="n">quantities</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">iy_ax</span><span class="p">,</span> <span class="n">qty</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">quantities</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">iy_ax</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="s1">&#39;beadsize&#39;</span><span class="p">]</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="n">qty</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;time (s)&#39;</span><span class="p">:</span> <span class="c1"># part 1 of adding second, minute and hour marker: plot the lines under the data</span>

        <span class="n">ynew</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">60</span><span class="p">,</span><span class="mi">600</span><span class="p">,</span><span class="mi">3600</span><span class="p">,</span><span class="mi">36000</span><span class="p">])</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">60</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">60</span><span class="p">,</span><span class="mi">600</span><span class="p">,</span><span class="mi">60</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">3600</span><span class="p">,</span><span class="mi">600</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">3600</span><span class="p">,</span><span class="mi">36000</span><span class="p">,</span><span class="mi">3600</span><span class="p">)])</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;0.1s&#39;</span><span class="p">,</span><span class="s1">&#39;1s&#39;</span><span class="p">,</span><span class="s1">&#39;10s&#39;</span><span class="p">,</span><span class="s1">&#39;1min&#39;</span><span class="p">,</span><span class="s1">&#39;10min&#39;</span><span class="p">,</span><span class="s1">&#39;1h&#39;</span><span class="p">,</span><span class="s1">&#39;10h&#39;</span><span class="p">])</span>
        <span class="n">ymin</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">*</span> <span class="mf">0.5</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">*</span> <span class="mf">2.0</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">ynewlabels</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew</span> <span class="o">=</span> <span class="n">ynew</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">ynew_minor</span><span class="p">[(</span><span class="n">ynew_minor</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew_minor</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">yn</span> <span class="ow">in</span> <span class="n">ynew</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">yn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;memory (GB)&#39;</span><span class="p">:</span>

        <span class="n">ynew</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">40</span><span class="p">,</span><span class="mi">100</span><span class="p">])</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="mi">100</span><span class="p">)])</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;0.1GB&#39;</span><span class="p">,</span><span class="s1">&#39;0.4GB&#39;</span><span class="p">,</span><span class="s1">&#39;1GB&#39;</span><span class="p">,</span><span class="s1">&#39;4GB&#39;</span><span class="p">,</span><span class="s1">&#39;10GB&#39;</span><span class="p">,</span><span class="s1">&#39;40GB&#39;</span><span class="p">,</span><span class="s1">&#39;100GB&#39;</span><span class="p">])</span>
        <span class="n">ymin</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">*</span> <span class="mf">0.5</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">*</span> <span class="mf">2.0</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">ynewlabels</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew</span> <span class="o">=</span> <span class="n">ynew</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">ynew_minor</span><span class="p">[(</span><span class="n">ynew_minor</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew_minor</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">yn</span> <span class="ow">in</span> <span class="n">ynew</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">yn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">:</span>
        <span class="n">selector</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="s1">&#39;method&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">m</span>
        <span class="k">if</span> <span class="n">selector</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">selector</span><span class="p">],</span><span class="n">y</span><span class="p">[</span><span class="n">selector</span><span class="p">],</span><span class="n">label</span><span class="o">=</span><span class="n">m</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">m</span><span class="p">],</span><span class="n">ls</span><span class="o">=</span><span class="n">styles</span><span class="p">[</span><span class="n">m</span><span class="p">],)</span>
    <span class="k">if</span> <span class="n">iy_ax</span> <span class="o">==</span> <span class="n">axs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;bead size&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;time (s)&#39;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;runtime&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;memory (GB)&#39;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;memory&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">qty</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">qty</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;time (s)&#39;</span><span class="p">,</span><span class="s1">&#39;memory (GB)&#39;</span><span class="p">]:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">qty</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;time (s)&#39;</span><span class="p">,</span><span class="s1">&#39;memory (GB)&#39;</span><span class="p">]:</span> <span class="c1"># part 2 off adding second, minute and hour marker: add the second y axis after rescaling the first y axis to log scale</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">ynew_minor</span><span class="p">,</span><span class="n">minor</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">ynew</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ynewlabels</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([],</span><span class="n">minor</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">iy_ax</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_insilico_mouse_colon_22_0.png" src="../_images/notebooks_insilico_mouse_colon_22_0.png" />
</div>
</div>
</section>
<section id="Observation-splitting">
<h2>Observation splitting<a class="headerlink" href="#Observation-splitting" title="Link to this heading">Â¶</a></h2>
<section id="Annotate-the-spatial-data-with-compositions-of-cell-types">
<h3>Annotate the spatial data with compositions of cell types<a class="headerlink" href="#Annotate-the-spatial-data-with-compositions-of-cell-types" title="Link to this heading">Â¶</a></h3>
<p>Annotation is done on cell type level with multi_center=10 to capture variation within a cell type and with reconstruction_key set to save the annotation on the level of subclustered cell types for distributing all counts to pure cell type observations later.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">tdata</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="s1">&#39;labels&#39;</span><span class="p">,</span> <span class="n">result_key</span><span class="o">=</span><span class="s1">&#39;TACCO&#39;</span><span class="p">,</span> <span class="n">multi_center</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">reconstruction_key</span><span class="o">=</span><span class="s1">&#39;rec&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting preprocessing
Annotation profiles were not found in `reference.varm[&#34;labels&#34;]`. Constructing reference profiles with `tacco.preprocessing.construct_reference_profiles` and default arguments...
Finished preprocessing in 11.76 seconds.
Starting annotation of data with shape (9971, 19661) and a reference of shape (17512, 19661) using the following wrapped method:
+- platform normalization: platform_iterations=0, gene_keys=labels, normalize_to=adata
   +- multi center: multi_center=10 multi_center_amplitudes=True
      +- bisection boost: bisections=4, bisection_divisor=3
         +- core: method=OT annotation_prior=None
mean,std( rescaling(gene) )  0.7484176896456394 0.24026163784656937
bisection run on 1
bisection run on 0.6666666666666667
bisection run on 0.4444444444444444
bisection run on 0.2962962962962963
bisection run on 0.19753086419753085
bisection run on 0.09876543209876543
Finished annotation in 68.32 seconds.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs Ã n_vars = 9971 Ã 31053
    obs: &#39;x&#39;, &#39;y&#39;
    var: &#39;gene_ids&#39;
    uns: &#39;rec&#39;
    obsm: &#39;labels&#39;, &#39;reads_labels&#39;, &#39;rec&#39;, &#39;TACCO&#39;
    varm: &#39;rec&#39;
</pre></div></div>
</div>
</section>
<section id="Split-the-mixed-observations-into-pure-contributions">
<h3>Split the mixed observations into pure contributions<a class="headerlink" href="#Split-the-mixed-observations-into-pure-contributions" title="Link to this heading">Â¶</a></h3>
<p>Here we use the annotation on the subcluster level, saved during annotation in the reconstruction_key.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sdata</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">split_observations</span><span class="p">(</span><span class="n">tdata</span><span class="p">,</span> <span class="s1">&#39;rec&#39;</span><span class="p">,</span> <span class="n">map_obs_keys</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">result_key</span><span class="o">=</span><span class="s1">&#39;labels&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
removed 11392 of 31053 genes from count matrix due to zero counts in gene
removed 11392 of 31053 genes from profile definition due to zero appearance in the profiles
scale.....time 142.23044776916504
fuseall...time 89.69051170349121
</pre></div></div>
</div>
</section>
<section id="Visualize-the-split-observations">
<h3>Visualize the split observations<a class="headerlink" href="#Visualize-the-split-observations" title="Link to this heading">Â¶</a></h3>
<p>Create UMAP for the in-silico mixtures after observation split</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">umap_sdata</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">umap_single_cell_data</span><span class="p">(</span><span class="n">sdata</span><span class="p">[</span><span class="n">tc</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">sdata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span><span class="o">&gt;=</span><span class="mi">500</span><span class="p">])</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">umap_sdata</span><span class="p">,</span> <span class="s1">&#39;labels&#39;</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">labels_colors</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">axes_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP 0&#39;</span><span class="p">,</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SCumap...SCprep...time 7.842227220535278
time 32.2546226978302
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_insilico_mouse_colon_32_1.png" src="../_images/notebooks_insilico_mouse_colon_32_1.png" />
</div>
</div>
<p>Create UMAP for a joint embedding of reference and observation-split in-silico mixtures</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">prep_ref</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">preprocess_single_cell_data</span><span class="p">(</span><span class="n">reference</span><span class="p">,</span> <span class="n">hvg</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">sdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;HVGcounts&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">sdata</span><span class="p">[:,</span><span class="n">sdata</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="n">prep_ref</span><span class="o">.</span><span class="n">var</span><span class="o">.</span><span class="n">index</span><span class="p">)]</span><span class="o">.</span><span class="n">X</span><span class="p">,</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">sdata</span> <span class="o">=</span> <span class="n">sdata</span><span class="p">[</span><span class="n">sdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;HVGcounts&#39;</span><span class="p">]</span><span class="o">&gt;=</span><span class="mi">50</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">scat</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">sdata</span><span class="p">],</span><span class="n">index_unique</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
<span class="n">umap_scat</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">umap_single_cell_data</span><span class="p">(</span><span class="n">scat</span><span class="p">)</span>
<span class="n">reference</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">umap_scat</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">umap_scat</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">sdata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">umap_scat</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;X_umap&#39;</span><span class="p">],</span><span class="n">index</span><span class="o">=</span><span class="n">umap_scat</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">sdata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>

<span class="n">fig</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">({</span><span class="s1">&#39;reference&#39;</span><span class="p">:</span><span class="n">reference</span><span class="p">,</span><span class="s1">&#39;split&#39;</span><span class="p">:</span><span class="n">sdata</span><span class="p">},</span> <span class="s1">&#39;labels&#39;</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">labels_colors</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">axes_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP 0&#39;</span><span class="p">,</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SCprep...time 8.178986072540283
SCumap...SCprep...time 20.207684755325317
time 75.30178141593933
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_insilico_mouse_colon_34_1.png" src="../_images/notebooks_insilico_mouse_colon_34_1.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="cooccurrence.html" class="btn btn-neutral float-left" title="Co-occurrence computations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="mapping_single_cells_into_space.html" class="btn btn-neutral float-right" title="Mapping single cells into space" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>