<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Simulated single cell expression data with dropout &mdash; TACCO  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=ad0ac74c" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
      <link rel="stylesheet" type="text/css" href="../_static/readthedocs-custom.css?v=c8300288" />
      <link rel="stylesheet" type="text/css" href="../_static/notebook_hacks.css?v=e4ce2130" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Single-cell differentiation" href="single_cell_differentiation.html" />
    <link rel="prev" title="Simulated single cell expression data with ambient contamination" href="simulated_ambient.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="benchmarking.html">Benchmarking runtime and memory of annotation methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="cooccurrence.html">Co-occurrence computations</a></li>
<li class="toctree-l2"><a class="reference internal" href="insilico_mouse_colon.html">In-silico mixed Mouse Colon scRNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="mapping_single_cells_into_space.html">Mapping single cells into space</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_splitting.html">Object splitting for in-silico mixed scRNA-seq data and Slide-Seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="osmFISH_single_molecule_annotation.html">Single molecule data: osmFISH mouse somato-sensory cortex</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html">TACCO overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="parameter_variations.html">Parameter Variations for annotation using Optimal Transport</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulated_ambient.html">Simulated single cell expression data with ambient contamination</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Simulated single cell expression data with dropout</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Simulate-single-cell-data">Simulate single cell data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plotting-options">Plotting options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Visualize-simulated-data-without-dropout">Visualize simulated data without dropout</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Visualize-simulated-data-with-dropout">Visualize simulated data with dropout</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Annotate-the-data-with-dropout-with-cell-types">Annotate the data with dropout with cell types</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Benchmark-annotation-methods-on-the-simulated-dropout-data">Benchmark annotation methods on the simulated dropout data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="single_cell_differentiation.html">Single-cell differentiation</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_colon.html">Slide-Seq Mouse Colon</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_olfactory_bulb.html">Slide-Seq Mouse Olfactory Bulb - multiple pucks</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_olfactory_bulb_single.html">Slide-Seq Mouse Olfactory Bulb - single puck</a></li>
<li class="toctree-l2"><a class="reference internal" href="visium_mouse_kidney.html">Visium Mouse Kidney</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../_autosummary/tacco.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TACCO</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../examples.html">Examples</a></li>
      <li class="breadcrumb-item active">Simulated single cell expression data with dropout</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Simulated-single-cell-expression-data-with-dropout">
<h1>Simulated single cell expression data with dropout<a class="headerlink" href="#Simulated-single-cell-expression-data-with-dropout" title="Link to this heading">¶</a></h1>
<p>This example uses TACCO to annotate simulated single cell data, simulated using a modified variant (Moriel) of scsim (Kotliar).</p>
<p>(Moriel): Moriel, N. Extension of scsim single-cell RNA-sequencing data simulations. github.com/nitzanlab/scsim-ext (2023)</p>
<p>(Kotliar): Kotliar, D. scsim: simulate single-cell RNA-SEQ data using the Splatter statistical framework but implemented in python. github.com/dylkot/scsim (2021)</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">matplotlib</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">scipy.sparse</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>

<span class="kn">import</span> <span class="nn">tacco</span> <span class="k">as</span> <span class="nn">tc</span>
<span class="kn">from</span> <span class="nn">scsim</span> <span class="kn">import</span> <span class="n">scsim</span>

<span class="c1"># The notebook expects to be executed either in the workflow directory or in the repository root folder...</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;workflow&#39;</span> <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;workflow/common_code.py&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;..&#39;</span><span class="p">))</span>
<span class="kn">import</span> <span class="nn">common_code</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">env_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/env_links&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Simulate-single-cell-data">
<h2>Simulate single cell data<a class="headerlink" href="#Simulate-single-cell-data" title="Link to this heading">¶</a></h2>
<p>Simulate data with various dropout levels with a known ground truth</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ngenes</span> <span class="o">=</span> <span class="mi">25000</span>
<span class="n">descale</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="n">ndoublets</span> <span class="o">=</span> <span class="mi">100</span>
<span class="n">K</span> <span class="o">=</span> <span class="mi">13</span>
<span class="n">nproggenes</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="n">proggroups</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]</span>
<span class="n">progcellfrac</span> <span class="o">=</span> <span class="mf">.35</span>
<span class="n">ncells</span> <span class="o">=</span> <span class="mi">1500</span>
<span class="n">deprob</span> <span class="o">=</span> <span class="mf">.025</span>

<span class="n">seed</span> <span class="o">=</span> <span class="mi">111</span>

<span class="n">deloc</span> <span class="o">=</span> <span class="mf">2.0</span>

<span class="c1"># simulating true counts (in simulator.counts)</span>
<span class="n">simulator</span> <span class="o">=</span> <span class="n">scsim</span><span class="p">(</span><span class="n">ngenes</span><span class="o">=</span><span class="n">ngenes</span><span class="p">,</span> <span class="n">ncells</span><span class="o">=</span><span class="n">ncells</span><span class="p">,</span> <span class="n">ngroups</span><span class="o">=</span><span class="n">K</span><span class="p">,</span> <span class="n">libloc</span><span class="o">=</span><span class="mf">7.64</span><span class="p">,</span> <span class="n">libscale</span><span class="o">=</span><span class="mf">0.78</span><span class="p">,</span>
             <span class="n">mean_rate</span><span class="o">=</span><span class="mf">7.68</span><span class="p">,</span><span class="n">mean_shape</span><span class="o">=</span><span class="mf">0.34</span><span class="p">,</span> <span class="n">expoutprob</span><span class="o">=</span><span class="mf">0.00286</span><span class="p">,</span>
             <span class="n">expoutloc</span><span class="o">=</span><span class="mf">6.15</span><span class="p">,</span> <span class="n">expoutscale</span><span class="o">=</span><span class="mf">0.49</span><span class="p">,</span>
             <span class="n">diffexpprob</span><span class="o">=</span><span class="n">deprob</span><span class="p">,</span> <span class="n">diffexpdownprob</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">diffexploc</span><span class="o">=</span><span class="n">deloc</span><span class="p">,</span> <span class="n">diffexpscale</span><span class="o">=</span><span class="n">descale</span><span class="p">,</span>
             <span class="n">bcv_dispersion</span><span class="o">=</span><span class="mf">0.448</span><span class="p">,</span> <span class="n">bcv_dof</span><span class="o">=</span><span class="mf">22.087</span><span class="p">,</span> <span class="n">ndoublets</span><span class="o">=</span><span class="n">ndoublets</span><span class="p">,</span>
             <span class="n">nproggenes</span><span class="o">=</span><span class="n">nproggenes</span><span class="p">,</span> <span class="n">progdownprob</span><span class="o">=</span><span class="mf">0.</span><span class="p">,</span> <span class="n">progdeloc</span><span class="o">=</span><span class="n">deloc</span><span class="p">,</span>
             <span class="n">progdescale</span><span class="o">=</span><span class="n">descale</span><span class="p">,</span> <span class="n">progcellfrac</span><span class="o">=</span><span class="n">progcellfrac</span><span class="p">,</span> <span class="n">proggoups</span><span class="o">=</span><span class="n">proggroups</span><span class="p">,</span>
             <span class="n">minprogusage</span><span class="o">=</span><span class="mf">.1</span><span class="p">,</span> <span class="n">maxprogusage</span><span class="o">=</span><span class="mf">.7</span><span class="p">,</span> <span class="n">seed</span><span class="o">=</span><span class="n">seed</span><span class="p">)</span>
<span class="n">simulator</span><span class="o">.</span><span class="n">simulate</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Simulating cells
Simulating gene params
Simulating program
Simulating DE
Simulating cell-gene means
   - Getting mean for activity program carrying cells
   - Getting mean for non activity program carrying cells
   - Normalizing by cell libsize
Simulating doublets
Adjusting means
Simulating counts with scsim
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">reference</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">simulator</span><span class="o">.</span><span class="n">counts</span><span class="p">),</span> <span class="n">obs</span><span class="o">=</span><span class="n">simulator</span><span class="o">.</span><span class="n">cellparams</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">simulator</span><span class="o">.</span><span class="n">geneparams</span><span class="p">)</span>
<span class="n">reference</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">reference</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>
<span class="c1"># fitting number of zeros vs log mean curve</span>
<span class="n">dropshape</span><span class="p">,</span> <span class="n">dropmidpoint</span> <span class="o">=</span> <span class="n">simulator</span><span class="o">.</span><span class="n">fit_dropout</span><span class="p">()</span>
<span class="n">dropout_midpoints</span> <span class="o">=</span> <span class="p">[</span><span class="mf">1.0</span><span class="p">,</span><span class="o">-</span><span class="mf">1.0</span><span class="p">,</span><span class="o">-</span><span class="mf">3.0</span><span class="p">,</span><span class="o">-</span><span class="mf">5.0</span><span class="p">,</span><span class="o">-</span><span class="mf">7.0</span><span class="p">]</span>

<span class="c1"># to add technical dropout noise, we can decrease the sigmoid&#39;s dropmidpoint to increase the dropout (in simulator.countswdrop)</span>
<span class="n">sdatas</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">dropout_midpoint</span> <span class="ow">in</span> <span class="n">dropout_midpoints</span><span class="p">:</span>
    <span class="n">simulator</span><span class="o">.</span><span class="n">dropshape</span> <span class="o">=</span> <span class="n">dropshape</span>
    <span class="n">simulator</span><span class="o">.</span><span class="n">dropmidpoint</span> <span class="o">=</span> <span class="n">dropout_midpoint</span>
    <span class="n">simulator</span><span class="o">.</span><span class="n">simulate_dropouts</span><span class="p">()</span>

    <span class="n">sdata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">AnnData</span><span class="p">(</span><span class="n">scipy</span><span class="o">.</span><span class="n">sparse</span><span class="o">.</span><span class="n">csr_matrix</span><span class="p">(</span><span class="n">simulator</span><span class="o">.</span><span class="n">countswdrop</span><span class="p">),</span> <span class="n">obs</span><span class="o">=</span><span class="n">simulator</span><span class="o">.</span><span class="n">cellparams</span><span class="p">,</span> <span class="n">var</span><span class="o">=</span><span class="n">simulator</span><span class="o">.</span><span class="n">geneparams</span><span class="p">)</span>
    <span class="n">sdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sdata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;group&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s1">&#39;category&#39;</span><span class="p">)</span>

    <span class="n">sdatas</span><span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">]</span> <span class="o">=</span> <span class="n">sdata</span>
</pre></div>
</div>
</div>
</section>
<section id="Plotting-options">
<h2>Plotting options<a class="headerlink" href="#Plotting-options" title="Link to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">highres</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">default_dpi</span> <span class="o">=</span> <span class="mf">100.0</span>
<span class="k">if</span> <span class="n">highres</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">648.0</span>
    <span class="n">hr_ext</span> <span class="o">=</span> <span class="s1">&#39;_hd&#39;</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">default_dpi</span>
    <span class="n">hr_ext</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

<span class="n">axsize</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span><span class="o">*</span><span class="mf">1.0</span>
</pre></div>
</div>
</div>
</section>
<section id="Visualize-simulated-data-without-dropout">
<h2>Visualize simulated data without dropout<a class="headerlink" href="#Visualize-simulated-data-without-dropout" title="Link to this heading">¶</a></h2>
<p>Create UMAP for the simulated data without dropout</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ref_umap</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">umap_single_cell_data</span><span class="p">(</span><span class="n">reference</span><span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">ref_umap</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">axes_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP 0&#39;</span><span class="p">,</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SCumap...SCprep...time 1.4392147064208984
time 26.896665811538696
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_simulated_dropout_12_1.png" src="../_images/notebooks_simulated_dropout_12_1.png" />
</div>
</div>
</section>
<section id="Visualize-simulated-data-with-dropout">
<h2>Visualize simulated data with dropout<a class="headerlink" href="#Visualize-simulated-data-with-dropout" title="Link to this heading">¶</a></h2>
<p>Create UMAP for the simulated data with dropout</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdata</span> <span class="o">=</span> <span class="n">sdatas</span><span class="p">[</span><span class="o">-</span><span class="mf">1.0</span><span class="p">]</span>
<span class="n">tdata_umap</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">umap_single_cell_data</span><span class="p">(</span><span class="n">tdata</span><span class="p">)</span>
<span class="n">fig</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">tdata_umap</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="n">axes_labels</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;UMAP 0&#39;</span><span class="p">,</span><span class="s1">&#39;UMAP 1&#39;</span><span class="p">]);</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
SCumap...SCprep...time 0.5789980888366699
time 8.304261207580566
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_simulated_dropout_15_1.png" src="../_images/notebooks_simulated_dropout_15_1.png" />
</div>
</div>
</section>
<section id="Annotate-the-data-with-dropout-with-cell-types">
<h2>Annotate the data with dropout with cell types<a class="headerlink" href="#Annotate-the-data-with-dropout-with-cell-types" title="Link to this heading">¶</a></h2>
<p>Annotation is done on cell type level with multi_center=10 to capture variation within a cell type.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">tdata</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">result_key</span><span class="o">=</span><span class="s1">&#39;TACCO&#39;</span><span class="p">,</span> <span class="n">multi_center</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">assume_valid_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Starting preprocessing
Annotation profiles were not found in `reference.varm[&#34;group&#34;]`. Constructing reference profiles with `tacco.preprocessing.construct_reference_profiles` and default arguments...
Finished preprocessing in 0.42 seconds.
Starting annotation of data with shape (1500, 19737) and a reference of shape (1500, 19737) using the following wrapped method:
+- platform normalization: platform_iterations=0, gene_keys=group, normalize_to=adata
   +- multi center: multi_center=10 multi_center_amplitudes=True
      +- bisection boost: bisections=4, bisection_divisor=3
         +- core: method=OT annotation_prior=None
mean,std( rescaling(gene) )  0.37671331182229645 0.19547104710406613
bisection run on 1
bisection run on 0.6666666666666667
bisection run on 0.4444444444444444
bisection run on 0.2962962962962963
bisection run on 0.19753086419753085
bisection run on 0.09876543209876543
Finished annotation in 7.44 seconds.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 1500 × 25000
    obs: &#39;group&#39;, &#39;libsize&#39;, &#39;has_program&#39;, &#39;program_usage&#39;, &#39;is_doublet&#39;, &#39;group2&#39;
    var: &#39;BaseGeneMean&#39;, &#39;is_outlier&#39;, &#39;outlier_ratio&#39;, &#39;gene_mean&#39;, &#39;prog_gene&#39;, &#39;prog_genemean&#39;, &#39;group1_DEratio&#39;, &#39;group1_genemean&#39;, &#39;group2_DEratio&#39;, &#39;group2_genemean&#39;, &#39;group3_DEratio&#39;, &#39;group3_genemean&#39;, &#39;group4_DEratio&#39;, &#39;group4_genemean&#39;, &#39;group5_DEratio&#39;, &#39;group5_genemean&#39;, &#39;group6_DEratio&#39;, &#39;group6_genemean&#39;, &#39;group7_DEratio&#39;, &#39;group7_genemean&#39;, &#39;group8_DEratio&#39;, &#39;group8_genemean&#39;, &#39;group9_DEratio&#39;, &#39;group9_genemean&#39;, &#39;group10_DEratio&#39;, &#39;group10_genemean&#39;, &#39;group11_DEratio&#39;, &#39;group11_genemean&#39;, &#39;group12_DEratio&#39;, &#39;group12_genemean&#39;, &#39;group13_DEratio&#39;, &#39;group13_genemean&#39;
    uns: &#39;TACCO_mc10&#39;
    obsm: &#39;TACCO_mc10&#39;, &#39;TACCO&#39;
    varm: &#39;TACCO_mc10&#39;
</pre></div></div>
</div>
</section>
<section id="Benchmark-annotation-methods-on-the-simulated-dropout-data">
<h2>Benchmark annotation methods on the simulated dropout data<a class="headerlink" href="#Benchmark-annotation-methods-on-the-simulated-dropout-data" title="Link to this heading">¶</a></h2>
<p>Define parameters for the annotation methods to use</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">methods</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;NMFreg&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;NMFreg&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;RCTD&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;RCTD&#39;</span><span class="p">,</span> <span class="s1">&#39;conda_env&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">env_path</span><span class="si">}</span><span class="s1">/RCTD_env&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;SVM&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span><span class="s1">&#39;svm&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;SingleR&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span><span class="s1">&#39;SingleR&#39;</span><span class="p">,</span> <span class="s1">&#39;conda_env&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">env_path</span><span class="si">}</span><span class="s1">/SingleR_env&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;WOT&#39;</span><span class="p">:{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;WOT&#39;</span><span class="p">,},</span>
    <span class="s1">&#39;TACCO&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;OT&#39;</span><span class="p">,</span> <span class="s1">&#39;multi_center&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,},</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">results</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">dropout_midpoint</span> <span class="ow">in</span> <span class="n">dropout_midpoints</span><span class="p">:</span>
    <span class="k">for</span> <span class="n">method</span><span class="p">,</span><span class="n">params</span> <span class="ow">in</span> <span class="n">methods</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;running method </span><span class="si">{</span><span class="n">method</span><span class="si">}</span><span class="s1"> for dropout midpoint </span><span class="si">{</span><span class="n">dropout_midpoint</span><span class="si">}</span><span class="s1"> ...&#39;</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">results</span><span class="p">[(</span><span class="n">dropout_midpoint</span><span class="p">,</span><span class="n">method</span><span class="p">)]</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">benchmarking</span><span class="o">.</span><span class="n">benchmark_annotate</span><span class="p">(</span><span class="n">sdatas</span><span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">],</span><span class="n">reference</span><span class="p">,</span><span class="n">annotation_key</span><span class="o">=</span><span class="s1">&#39;group&#39;</span><span class="p">,</span><span class="o">**</span><span class="n">params</span><span class="p">,</span> <span class="n">assume_valid_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="k">pass</span> <span class="c1"># catch failing methods</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;done&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
running method NMFreg for dropout midpoint 1.0 ...done
running method RCTD for dropout midpoint 1.0 ...done
running method SVM for dropout midpoint 1.0 ...done
running method SingleR for dropout midpoint 1.0 ...done
running method WOT for dropout midpoint 1.0 ...done
running method TACCO for dropout midpoint 1.0 ...done
running method NMFreg for dropout midpoint -1.0 ...done
running method RCTD for dropout midpoint -1.0 ...done
running method SVM for dropout midpoint -1.0 ...done
running method SingleR for dropout midpoint -1.0 ...done
running method WOT for dropout midpoint -1.0 ...done
running method TACCO for dropout midpoint -1.0 ...done
running method NMFreg for dropout midpoint -3.0 ...done
running method RCTD for dropout midpoint -3.0 ...done
running method SVM for dropout midpoint -3.0 ...done
running method SingleR for dropout midpoint -3.0 ...done
running method WOT for dropout midpoint -3.0 ...done
running method TACCO for dropout midpoint -3.0 ...done
running method NMFreg for dropout midpoint -5.0 ...done
running method RCTD for dropout midpoint -5.0 ...done
running method SVM for dropout midpoint -5.0 ...done
running method SingleR for dropout midpoint -5.0 ...done
running method WOT for dropout midpoint -5.0 ...Starting preprocessing
Finished preprocessing in 0.39 seconds.
Starting annotation of data with shape (1500, 13619) and a reference of shape (1500, 13619) using the following wrapped method:
+- core: method=WOT annotation_prior=None
Exception in WaddingtonOT. Trying with cell normalization.

/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates(&#34;obs&#34;)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:134: RuntimeWarning: divide by zero encountered in true_divide
  b = (q / (K.T.dot(np.multiply(a, dx)))) ** alpha2 * np.exp(-v / (lambda2 + epsilon_i))
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:138: RuntimeWarning: divide by zero encountered in log
  u = u + epsilon_i * np.log(a)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:140: RuntimeWarning: invalid value encountered in add
  K = np.exp((np.array([u]).T - C + np.array([v])) / epsilon_i)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates(&#34;obs&#34;)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:159: RuntimeWarning: invalid value encountered in double_scalars
  np.linalg.norm(_a - old_a * np.exp(u / epsilon_i)) / (1 + np.linalg.norm(_a)),
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:148: RuntimeWarning: overflow encountered in exp
  _a = a * np.exp(u / epsilon_i)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:159: RuntimeWarning: overflow encountered in exp
  np.linalg.norm(_a - old_a * np.exp(u / epsilon_i)) / (1 + np.linalg.norm(_a)),
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:159: RuntimeWarning: invalid value encountered in subtract
  np.linalg.norm(_a - old_a * np.exp(u / epsilon_i)) / (1 + np.linalg.norm(_a)),
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:133: RuntimeWarning: overflow encountered in true_divide
  a = (p / (K.dot(np.multiply(b, dy)))) ** alpha1 * np.exp(-u / (lambda1 + epsilon_i))
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:139: RuntimeWarning: divide by zero encountered in log
  v = v + epsilon_i * np.log(b)  # absorb
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:140: RuntimeWarning: invalid value encountered in add
  K = np.exp((np.array([u]).T - C + np.array([v])) / epsilon_i)
Traceback (most recent call last):
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_wot.py&#34;, line 42, in _annotate_wot
    tmap_annotated = ot_model.compute_transport_map(0, 1)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 231, in compute_transport_map
    return self.compute_single_transport_map(config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 304, in compute_single_transport_map
    tmap, learned_growth = wot.ot.compute_transport_matrix(solver=self.solver, **config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 30, in compute_transport_matrix
    tmap = solver(**params)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 163, in optimal_transport_duality_gap
    raise RuntimeError(&#34;Overflow encountered in duality gap computation, please report this incident&#34;)
RuntimeError: Overflow encountered in duality gap computation, please report this incident

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/workflow/simulated_dropout/temp_benchmark_sntgs8wj/benchmark_script_bxfknfzf.py&#34;, line 22, in &lt;module&gt;
    result = tc.tl.annotate(adata, reference, **kw_args)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_annotate.py&#34;, line 802, in annotate
    cell_type = _method(tdata, reference, annotation_key, annotation_prior, verbose)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_annotate.py&#34;, line 119, in _method
    cell_type = annotate(adata, reference, annotation_key, **verbose_arg, **kw_args)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_wot.py&#34;, line 47, in _annotate_wot
    tmap_annotated = ot_model.compute_transport_map(0, 1)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 231, in compute_transport_map
    return self.compute_single_transport_map(config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 304, in compute_single_transport_map
    tmap, learned_growth = wot.ot.compute_transport_matrix(solver=self.solver, **config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 30, in compute_transport_matrix
    tmap = solver(**params)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 163, in optimal_transport_duality_gap
    raise RuntimeError(&#34;Overflow encountered in duality gap computation, please report this incident&#34;)
RuntimeError: Overflow encountered in duality gap computation, please report this incident
wall_clock_time_seconds 12.60
max_memory_used_kbytes 1066564
exit_status 1

done
running method TACCO for dropout midpoint -5.0 ...done
running method NMFreg for dropout midpoint -7.0 ...done
running method RCTD for dropout midpoint -7.0 ...Starting preprocessing
Finished preprocessing in 0.33 seconds.
Starting annotation of data with shape (1483, 6771) and a reference of shape (1500, 6771) using the following wrapped method:
+- core: method=RCTD annotation_prior=None conda_env=/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/results/env_links/RCTD_env
 [1] &#34;reference.h5ad&#34; &#34;data.h5ad&#34;      &#34;result&#34;         &#34;0&#34;
 [5] &#34;8&#34;              &#34;full&#34;           &#34;group&#34;          &#34;x&#34;
 [9] &#34;y&#34;              &#34;7.0&#34;
[1] &#34;reading data&#34;
[1] &#34;csr&#34;
[1] &#34;reading reference&#34;
[1] &#34;csr&#34;
[1] &#34;running RCTD&#34;

  1   2   3   4   5   6   7   8   9  10  11  12  13
110 122 102 112 112 132 110 122 108 116 120 112 121


Attaching package: &#39;SparseM&#39;

The following object is masked from &#39;package:base&#39;:

    backsolve

Begin: process_cell_type_info
process_cell_type_info: number of cells in reference: 1499
process_cell_type_info: number of genes in reference: 6771
End: process_cell_type_info
create.RCTD: getting regression differentially expressed genes:
get_de_genes: 1 found DE genes: 31
get_de_genes: 2 found DE genes: 25
get_de_genes: 3 found DE genes: 26
get_de_genes: 4 found DE genes: 18
get_de_genes: 5 found DE genes: 25
get_de_genes: 6 found DE genes: 22
get_de_genes: 7 found DE genes: 16
get_de_genes: 8 found DE genes: 23
get_de_genes: 9 found DE genes: 29
get_de_genes: 10 found DE genes: 18
get_de_genes: 11 found DE genes: 18
get_de_genes: 12 found DE genes: 19
get_de_genes: 13 found DE genes: 16
get_de_genes: total DE genes: 260
create.RCTD: getting platform effect normalization differentially expressed genes:
get_de_genes: 1 found DE genes: 54
get_de_genes: 2 found DE genes: 42
get_de_genes: 3 found DE genes: 62
get_de_genes: 4 found DE genes: 40
get_de_genes: 5 found DE genes: 56
get_de_genes: 6 found DE genes: 44
get_de_genes: 7 found DE genes: 35
get_de_genes: 8 found DE genes: 48
get_de_genes: 9 found DE genes: 49
get_de_genes: 10 found DE genes: 48
get_de_genes: 11 found DE genes: 43
get_de_genes: 12 found DE genes: 28
get_de_genes: 13 found DE genes: 38
get_de_genes: total DE genes: 440
fitBulk: decomposing bulk
Error in choose_sigma_c(RCTD) :
  choose_sigma_c determined a N_fit of 0! This is probably due to unusually low UMI counts per bead in your dataset. Try decreasing the parameter UMI_min_sigma. It currently is 7.0 but none of the beads had counts larger than that.
Calls: run.RCTD -&gt; choose_sigma_c
Execution halted


/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_RCTD.py:239: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.
  adata.obs[x_coord_name] = 0
Traceback (most recent call last):
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/workflow/simulated_dropout/temp_benchmark_yz4_fh5y/benchmark_script_jq4d9rx5.py&#34;, line 22, in &lt;module&gt;
    result = tc.tl.annotate(adata, reference, **kw_args)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_annotate.py&#34;, line 802, in annotate
    cell_type = _method(tdata, reference, annotation_key, annotation_prior, verbose)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_annotate.py&#34;, line 119, in _method
    cell_type = annotate(adata, reference, annotation_key, **verbose_arg, **kw_args)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_RCTD.py&#34;, line 273, in annotate_RCTD
    raise Exception(&#39;RCTD did not work properly!&#39;)
Exception: RCTD did not work properly!
wall_clock_time_seconds 18.20
max_memory_used_kbytes 510112
exit_status 1

done
running method SVM for dropout midpoint -7.0 ...done
running method SingleR for dropout midpoint -7.0 ...done
running method WOT for dropout midpoint -7.0 ...Starting preprocessing
Finished preprocessing in 0.33 seconds.
Starting annotation of data with shape (1483, 6771) and a reference of shape (1500, 6771) using the following wrapped method:
+- core: method=WOT annotation_prior=None
Exception in WaddingtonOT. Trying with cell normalization.

/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates(&#34;obs&#34;)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:134: RuntimeWarning: divide by zero encountered in true_divide
  b = (q / (K.T.dot(np.multiply(a, dx)))) ** alpha2 * np.exp(-v / (lambda2 + epsilon_i))
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:138: RuntimeWarning: divide by zero encountered in log
  u = u + epsilon_i * np.log(a)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:140: RuntimeWarning: invalid value encountered in add
  K = np.exp((np.array([u]).T - C + np.array([v])) / epsilon_i)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/anndata/_core/anndata.py:1828: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates(&#34;obs&#34;)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:159: RuntimeWarning: invalid value encountered in double_scalars
  np.linalg.norm(_a - old_a * np.exp(u / epsilon_i)) / (1 + np.linalg.norm(_a)),
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:148: RuntimeWarning: overflow encountered in exp
  _a = a * np.exp(u / epsilon_i)
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:159: RuntimeWarning: overflow encountered in exp
  np.linalg.norm(_a - old_a * np.exp(u / epsilon_i)) / (1 + np.linalg.norm(_a)),
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:159: RuntimeWarning: invalid value encountered in subtract
  np.linalg.norm(_a - old_a * np.exp(u / epsilon_i)) / (1 + np.linalg.norm(_a)),
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:133: RuntimeWarning: divide by zero encountered in true_divide
  a = (p / (K.dot(np.multiply(b, dy)))) ** alpha1 * np.exp(-u / (lambda1 + epsilon_i))
/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py:133: RuntimeWarning: overflow encountered in true_divide
  a = (p / (K.dot(np.multiply(b, dy)))) ** alpha1 * np.exp(-u / (lambda1 + epsilon_i))
Traceback (most recent call last):
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_wot.py&#34;, line 42, in _annotate_wot
    tmap_annotated = ot_model.compute_transport_map(0, 1)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 231, in compute_transport_map
    return self.compute_single_transport_map(config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 304, in compute_single_transport_map
    tmap, learned_growth = wot.ot.compute_transport_matrix(solver=self.solver, **config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 30, in compute_transport_matrix
    tmap = solver(**params)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 163, in optimal_transport_duality_gap
    raise RuntimeError(&#34;Overflow encountered in duality gap computation, please report this incident&#34;)
RuntimeError: Overflow encountered in duality gap computation, please report this incident

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/workflow/simulated_dropout/temp_benchmark_53yspg0j/benchmark_script_k4mku6of.py&#34;, line 22, in &lt;module&gt;
    result = tc.tl.annotate(adata, reference, **kw_args)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_annotate.py&#34;, line 802, in annotate
    cell_type = _method(tdata, reference, annotation_key, annotation_prior, verbose)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_annotate.py&#34;, line 119, in _method
    cell_type = annotate(adata, reference, annotation_key, **verbose_arg, **kw_args)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco/tacco/tools/_wot.py&#34;, line 47, in _annotate_wot
    tmap_annotated = ot_model.compute_transport_map(0, 1)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 231, in compute_transport_map
    return self.compute_single_transport_map(config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/ot_model.py&#34;, line 304, in compute_single_transport_map
    tmap, learned_growth = wot.ot.compute_transport_matrix(solver=self.solver, **config)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 30, in compute_transport_matrix
    tmap = solver(**params)
  File &#34;/ahg/regevdata/projects/mouse_CRC/rerun/tacco_examples_devel/.snakemake/conda/1d2a210995279fd3d78066142f3cf09b/lib/python3.9/site-packages/wot/ot/optimal_transport.py&#34;, line 163, in optimal_transport_duality_gap
    raise RuntimeError(&#34;Overflow encountered in duality gap computation, please report this incident&#34;)
RuntimeError: Overflow encountered in duality gap computation, please report this incident
wall_clock_time_seconds 10.58
max_memory_used_kbytes 725876
exit_status 1

done
running method TACCO for dropout midpoint -7.0 ...done
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="p">(</span><span class="n">dropout_midpoint</span><span class="p">,</span><span class="n">method</span><span class="p">),</span><span class="n">result</span> <span class="ow">in</span> <span class="n">results</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="n">unused_key</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">find_unused_key</span><span class="p">(</span><span class="n">sdatas</span><span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">)</span>
    <span class="n">sdatas</span><span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">unused_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">results</span><span class="p">[(</span><span class="n">dropout_midpoint</span><span class="p">,</span><span class="n">method</span><span class="p">)][</span><span class="s1">&#39;annotation&#39;</span><span class="p">]</span>
    <span class="n">max_correct</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">ev</span><span class="o">.</span><span class="n">compute_err</span><span class="p">(</span><span class="n">sdatas</span><span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">],</span> <span class="n">unused_key</span><span class="p">,</span> <span class="s1">&#39;group&#39;</span><span class="p">,</span> <span class="n">err_method</span><span class="o">=</span><span class="s1">&#39;max_correct&#39;</span><span class="p">)[</span><span class="n">unused_key</span><span class="p">]</span>
    <span class="k">del</span> <span class="n">sdatas</span><span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">]</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">unused_key</span><span class="p">]</span>
    <span class="n">results</span><span class="p">[(</span><span class="n">dropout_midpoint</span><span class="p">,</span><span class="n">method</span><span class="p">)][</span><span class="s1">&#39;max_correct&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">max_correct</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">res_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([</span>
    <span class="p">[</span><span class="n">dropout_midpoint</span><span class="p">,</span><span class="n">method</span><span class="p">,</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;max_correct&#39;</span><span class="p">],</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;max_mem_usage_GB&#39;</span><span class="p">],</span><span class="n">v</span><span class="p">[</span><span class="s1">&#39;benchmark_time_s&#39;</span><span class="p">]]</span>
    <span class="k">for</span> <span class="p">(</span><span class="n">dropout_midpoint</span><span class="p">,</span><span class="n">method</span><span class="p">),</span><span class="n">v</span> <span class="ow">in</span> <span class="n">results</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
<span class="p">],</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dropout_midpoint&#39;</span><span class="p">,</span><span class="s1">&#39;method&#39;</span><span class="p">,</span><span class="s1">&#39;max_correct&#39;</span><span class="p">,</span><span class="s1">&#39;memory (GB)&#39;</span><span class="p">,</span><span class="s1">&#39;time (s)&#39;</span><span class="p">])</span>
<span class="n">quantities</span> <span class="o">=</span> <span class="p">[</span><span class="n">c</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="n">res_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">c</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;dropout_midpoint&#39;</span><span class="p">,</span><span class="s1">&#39;method&#39;</span><span class="p">]</span> <span class="p">]</span>
<span class="n">methods</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="s1">&#39;method&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">axs</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">quantities</span><span class="p">),</span> <span class="n">axsize</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span><span class="o">*</span><span class="mf">0.4</span><span class="p">,</span> <span class="n">x_padding</span><span class="o">=</span><span class="mf">0.7</span><span class="p">,</span> <span class="n">y_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="p">:</span><span class="n">common_code</span><span class="o">.</span><span class="n">method_color</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">}</span>
<span class="n">styles</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span><span class="p">:</span><span class="n">common_code</span><span class="o">.</span><span class="n">method_style</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">}</span>
<span class="n">res_df</span> <span class="o">=</span> <span class="n">res_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="o">~</span><span class="n">res_df</span><span class="p">[</span><span class="n">quantities</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()</span><span class="o">.</span><span class="n">all</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span>
<span class="k">for</span> <span class="n">iy_ax</span><span class="p">,</span> <span class="n">qty</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">quantities</span><span class="p">):</span>
    <span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="n">iy_ax</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>

    <span class="n">x</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="s1">&#39;dropout_midpoint&#39;</span><span class="p">]</span>
    <span class="n">y</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="n">qty</span><span class="p">]</span>

    <span class="k">if</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;time (s)&#39;</span><span class="p">:</span> <span class="c1"># part 1 of adding second, minute and hour marker: plot the lines under the data</span>

        <span class="n">ynew</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">60</span><span class="p">,</span><span class="mi">600</span><span class="p">,</span><span class="mi">3600</span><span class="p">,</span><span class="mi">36000</span><span class="p">])</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">60</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">60</span><span class="p">,</span><span class="mi">600</span><span class="p">,</span><span class="mi">60</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">600</span><span class="p">,</span><span class="mi">3600</span><span class="p">,</span><span class="mi">600</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">3600</span><span class="p">,</span><span class="mi">36000</span><span class="p">,</span><span class="mi">3600</span><span class="p">)])</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;0.1s&#39;</span><span class="p">,</span><span class="s1">&#39;1s&#39;</span><span class="p">,</span><span class="s1">&#39;10s&#39;</span><span class="p">,</span><span class="s1">&#39;1min&#39;</span><span class="p">,</span><span class="s1">&#39;10min&#39;</span><span class="p">,</span><span class="s1">&#39;1h&#39;</span><span class="p">,</span><span class="s1">&#39;10h&#39;</span><span class="p">])</span>
        <span class="n">ymin</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">*</span> <span class="mf">0.5</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">*</span> <span class="mf">2.0</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">ynewlabels</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew</span> <span class="o">=</span> <span class="n">ynew</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">ynew_minor</span><span class="p">[(</span><span class="n">ynew_minor</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew_minor</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">yn</span> <span class="ow">in</span> <span class="n">ynew</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">yn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

    <span class="k">elif</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;memory (GB)&#39;</span><span class="p">:</span>

        <span class="n">ynew</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mf">0.1</span><span class="p">,</span><span class="mf">0.4</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">40</span><span class="p">,</span><span class="mi">100</span><span class="p">])</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">concatenate</span><span class="p">([</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mf">0.1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mf">0.1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">10</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span><span class="mi">100</span><span class="p">,</span><span class="mi">10</span><span class="p">),</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">100</span><span class="p">,</span><span class="mi">1000</span><span class="p">,</span><span class="mi">100</span><span class="p">)])</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;0.1GB&#39;</span><span class="p">,</span><span class="s1">&#39;0.4GB&#39;</span><span class="p">,</span><span class="s1">&#39;1GB&#39;</span><span class="p">,</span><span class="s1">&#39;4GB&#39;</span><span class="p">,</span><span class="s1">&#39;10GB&#39;</span><span class="p">,</span><span class="s1">&#39;40GB&#39;</span><span class="p">,</span><span class="s1">&#39;100GB&#39;</span><span class="p">])</span>
        <span class="n">ymin</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">min</span><span class="p">()</span> <span class="o">*</span> <span class="mf">0.5</span>
        <span class="n">ymax</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">max</span><span class="p">()</span> <span class="o">*</span> <span class="mf">2.0</span>
        <span class="n">ynewlabels</span> <span class="o">=</span> <span class="n">ynewlabels</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew</span> <span class="o">=</span> <span class="n">ynew</span><span class="p">[(</span><span class="n">ynew</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="n">ynew_minor</span> <span class="o">=</span> <span class="n">ynew_minor</span><span class="p">[(</span><span class="n">ynew_minor</span> <span class="o">&gt;</span> <span class="n">ymin</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">ynew_minor</span> <span class="o">&lt;</span> <span class="n">ymax</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">yn</span> <span class="ow">in</span> <span class="n">ynew</span><span class="p">:</span>
            <span class="n">ax</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span><span class="n">yn</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;gray&#39;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">methods</span><span class="p">:</span>
        <span class="n">selector</span> <span class="o">=</span> <span class="n">res_df</span><span class="p">[</span><span class="s1">&#39;method&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">m</span>
        <span class="k">if</span> <span class="n">selector</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="n">selector</span><span class="p">],</span><span class="n">y</span><span class="p">[</span><span class="n">selector</span><span class="p">],</span><span class="n">label</span><span class="o">=</span><span class="n">m</span><span class="p">,</span><span class="n">marker</span><span class="o">=</span><span class="s1">&#39;o&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">[</span><span class="n">m</span><span class="p">],</span><span class="n">ls</span><span class="o">=</span><span class="n">styles</span><span class="p">[</span><span class="n">m</span><span class="p">],)</span>
    <span class="k">if</span> <span class="n">iy_ax</span> <span class="o">==</span> <span class="n">axs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s1">&#39;bead size&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;time (s)&#39;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;runtime&#39;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">qty</span> <span class="o">==</span> <span class="s1">&#39;memory (GB)&#39;</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;memory&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">qty</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">qty</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;time (s)&#39;</span><span class="p">,</span><span class="s1">&#39;memory (GB)&#39;</span><span class="p">]:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yscale</span><span class="p">(</span><span class="s1">&#39;log&#39;</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">qty</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;time (s)&#39;</span><span class="p">,</span><span class="s1">&#39;memory (GB)&#39;</span><span class="p">]:</span> <span class="c1"># part 2 off adding second, minute and hour marker: add the second y axis after rescaling the first y axis to log scale</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">ynew_minor</span><span class="p">,</span><span class="n">minor</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticks</span><span class="p">(</span><span class="n">ynew</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">(</span><span class="n">ynewlabels</span><span class="p">)</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">set_yticklabels</span><span class="p">([],</span><span class="n">minor</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="k">if</span> <span class="n">iy_ax</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_simulated_dropout_25_0.png" src="../_images/notebooks_simulated_dropout_25_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="simulated_ambient.html" class="btn btn-neutral float-left" title="Simulated single cell expression data with ambient contamination" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="single_cell_differentiation.html" class="btn btn-neutral float-right" title="Single-cell differentiation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>