<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Single-cell differentiation &mdash; TACCO  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=ad0ac74c" />
      <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css" />
      <link rel="stylesheet" type="text/css" href="../_static/readthedocs-custom.css?v=c8300288" />
      <link rel="stylesheet" type="text/css" href="../_static/notebook_hacks.css?v=e4ce2130" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/documentation_options.js?v=5929fcd5"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://cdn.jsdelivr.net/npm/@jupyter-widgets/html-manager@^1.0.1/dist/embed-amd.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Slide-Seq Mouse Colon" href="slideseq_mouse_colon.html" />
    <link rel="prev" title="Simulated single cell expression data with dropout" href="simulated_dropout.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html">
            
              <img src="../_static/logo.svg" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../index.html">Home</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../examples.html">Examples</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="benchmarking.html">Benchmarking runtime and memory of annotation methods</a></li>
<li class="toctree-l2"><a class="reference internal" href="cooccurrence.html">Co-occurrence computations</a></li>
<li class="toctree-l2"><a class="reference internal" href="insilico_mouse_colon.html">In-silico mixed Mouse Colon scRNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="mapping_single_cells_into_space.html">Mapping single cells into space</a></li>
<li class="toctree-l2"><a class="reference internal" href="object_splitting.html">Object splitting for in-silico mixed scRNA-seq data and Slide-Seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="osmFISH_single_molecule_annotation.html">Single molecule data: osmFISH mouse somato-sensory cortex</a></li>
<li class="toctree-l2"><a class="reference internal" href="overview.html">TACCO overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="parameter_variations.html">Parameter Variations for annotation using Optimal Transport</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulated_ambient.html">Simulated single cell expression data with ambient contamination</a></li>
<li class="toctree-l2"><a class="reference internal" href="simulated_dropout.html">Simulated single cell expression data with dropout</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Single-cell differentiation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Load-data">Load data</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plotting-options">Plotting options</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Run-fate-prediction">Run fate prediction</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Plot-results">Plot results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_colon.html">Slide-Seq Mouse Colon</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_olfactory_bulb.html">Slide-Seq Mouse Olfactory Bulb - multiple pucks</a></li>
<li class="toctree-l2"><a class="reference internal" href="slideseq_mouse_olfactory_bulb_single.html">Slide-Seq Mouse Olfactory Bulb - single puck</a></li>
<li class="toctree-l2"><a class="reference internal" href="visium_mouse_kidney.html">Visium Mouse Kidney</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../_autosummary/tacco.html">API reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="../release_notes.html">Release Notes</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">TACCO</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../examples.html">Examples</a></li>
      <li class="breadcrumb-item active">Single-cell differentiation</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Single-cell-differentiation">
<h1>Single-cell differentiation<a class="headerlink" href="#Single-cell-differentiation" title="Link to this heading">¶</a></h1>
<p>This example uses TACCO to predict the fate of single cells in a hematopoietic setting comparing against experimental ground truth (Weinreb et al.) and other methods.</p>
<p>(Weinreb et al.): Weinreb C, Rodriguez-Fraticelli A, Camargo FD, Klein AM. Lineage tracing on transcriptional landscapes links state to fate during differentiation. Science. 2020 Feb 14;367(6479):eaaw3381. doi: 10.1126/science.aaw3381. Epub 2020 Jan 23. PMID: 31974159; PMCID: PMC7608074.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">sys</span>

<span class="kn">import</span> <span class="nn">matplotlib</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>

<span class="kn">import</span> <span class="nn">tacco</span> <span class="k">as</span> <span class="nn">tc</span>

<span class="c1"># The notebook expects to be executed either in the workflow directory or in the repository root folder...</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">insert</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="s1">&#39;workflow&#39;</span> <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;workflow/common_code.py&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="s1">&#39;..&#39;</span><span class="p">))</span>
<span class="kn">import</span> <span class="nn">common_code</span>
</pre></div>
</div>
</div>
<section id="Load-data">
<h2>Load data<a class="headerlink" href="#Load-data" title="Link to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">data_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/single_cell_differentiation/data&#39;</span><span class="p">)</span>
<span class="n">plot_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/single_cell_differentiation&#39;</span><span class="p">)</span>
<span class="n">env_path</span> <span class="o">=</span> <span class="n">common_code</span><span class="o">.</span><span class="n">find_path</span><span class="p">(</span><span class="s1">&#39;results/env_links&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">d4d6</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s1">/d4_d6_differentiation.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">d2</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_path</span><span class="si">}</span><span class="s1">/d2_differentiation.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Plotting-options">
<h2>Plotting options<a class="headerlink" href="#Plotting-options" title="Link to this heading">¶</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">highres</span> <span class="o">=</span> <span class="kc">False</span>
<span class="n">default_dpi</span> <span class="o">=</span> <span class="mf">100.0</span> <span class="c1"># matplotlib.rcParams[&#39;figure.dpi&#39;]</span>
<span class="k">if</span> <span class="n">highres</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">648.0</span>
    <span class="n">hr_ext</span> <span class="o">=</span> <span class="s1">&#39;_hd&#39;</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">matplotlib</span><span class="o">.</span><span class="n">rcParams</span><span class="p">[</span><span class="s1">&#39;figure.dpi&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">default_dpi</span>
    <span class="n">hr_ext</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>

<span class="n">axsize</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span><span class="o">*</span><span class="mf">0.5</span>

<span class="n">x</span> <span class="o">=</span> <span class="s1">&#39;SPRING-x&#39;</span>
<span class="n">y</span> <span class="o">=</span> <span class="s1">&#39;SPRING-y&#39;</span>

<span class="n">dummy_val</span> <span class="o">=</span> <span class="s1">&#39;dummy&#39;</span>
<span class="n">only_gray</span> <span class="o">=</span> <span class="p">{</span><span class="n">dummy_val</span><span class="p">:</span><span class="s1">&#39;#CCCCCC&#39;</span><span class="p">}</span>
<span class="n">color_dict</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;Neutrophil&#39;</span><span class="p">:</span> <span class="s1">&#39;#e9ff00&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Erythroid&#39;</span><span class="p">:</span> <span class="s1">&#39;#5a00ff&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Monocyte&#39;</span><span class="p">:</span> <span class="s1">&#39;#0039ff&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Meg&#39;</span><span class="p">:</span> <span class="s1">&#39;#00ff00&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Mast&#39;</span><span class="p">:</span> <span class="s1">&#39;#ff00a3&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Baso&#39;</span><span class="p">:</span> <span class="s1">&#39;#ff8100&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Lymphoid&#39;</span><span class="p">:</span> <span class="s1">&#39;#00caff&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Eos&#39;</span><span class="p">:</span> <span class="s1">&#39;#eb00ff&#39;</span><span class="p">,</span>
    <span class="s1">&#39;Ccr7_DC&#39;</span><span class="p">:</span> <span class="s1">&#39;#00ff00&#39;</span><span class="p">,</span>
    <span class="s1">&#39;pDC&#39;</span><span class="p">:</span> <span class="s1">&#39;#00faa4&#39;</span><span class="p">,</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</section>
<section id="Run-fate-prediction">
<h2>Run fate prediction<a class="headerlink" href="#Run-fate-prediction" title="Link to this heading">¶</a></h2>
<p>Use a set of different methods for comparison</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">method_configs</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;TACCO&#39;</span><span class="p">:</span>   <span class="p">{</span> <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;OT&#39;</span><span class="p">,</span> <span class="s1">&#39;multi_center&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span> <span class="p">},</span> <span class="c1"># multi_center=10 takes into account variation within a single class</span>
    <span class="s1">&#39;WOT&#39;</span><span class="p">:</span>     <span class="p">{</span> <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;WOT&#39;</span><span class="p">,</span> <span class="p">},</span>
    <span class="s1">&#39;SVM&#39;</span><span class="p">:</span>     <span class="p">{</span> <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;svm&#39;</span><span class="p">,</span> <span class="p">},</span>
    <span class="s1">&#39;NMFreg&#39;</span><span class="p">:</span>  <span class="p">{</span> <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;NMFreg&#39;</span><span class="p">,</span> <span class="p">},</span>
    <span class="s1">&#39;SingleR&#39;</span><span class="p">:</span> <span class="p">{</span> <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;SingleR&#39;</span><span class="p">,</span> <span class="s1">&#39;conda_env&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">env_path</span><span class="si">}</span><span class="s1">/SingleR_env&#39;</span><span class="p">,</span> <span class="p">},</span>
    <span class="s1">&#39;RCTD&#39;</span><span class="p">:</span>    <span class="p">{</span> <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="s1">&#39;RCTD&#39;</span><span class="p">,</span> <span class="s1">&#39;conda_env&#39;</span><span class="p">:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">env_path</span><span class="si">}</span><span class="s1">/RCTD_env&#39;</span><span class="p">,</span> <span class="p">},</span>
<span class="p">}</span>
<span class="n">benchmark_results</span> <span class="o">=</span> <span class="p">{}</span>
<span class="k">for</span> <span class="n">method</span><span class="p">,</span> <span class="n">params</span> <span class="ow">in</span> <span class="n">method_configs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;running method </span><span class="si">{</span><span class="n">method</span><span class="si">!r}</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">annotate</span><span class="p">(</span><span class="n">d2</span><span class="p">,</span> <span class="n">d4d6</span><span class="p">,</span> <span class="n">annotation_key</span><span class="o">=</span><span class="s1">&#39;Cell type annotation&#39;</span><span class="p">,</span> <span class="n">result_key</span><span class="o">=</span><span class="n">method</span><span class="p">,</span> <span class="o">**</span><span class="n">method_configs</span><span class="p">[</span><span class="n">method</span><span class="p">],</span> <span class="n">assume_valid_counts</span><span class="o">=</span><span class="kc">True</span><span class="p">,)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">close</span><span class="p">(</span><span class="s2">&quot;all&quot;</span><span class="p">)</span> <span class="c1"># suppress side-effect plots from NMFreg</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
running method &#39;TACCO&#39;
Starting preprocessing
Annotation profiles were not found in `reference.varm[&#34;Cell type annotation&#34;]`. Constructing reference profiles with `tacco.preprocessing.construct_reference_profiles` and default arguments...
Finished preprocessing in 2.26 seconds.
Starting annotation of data with shape (976, 16624) and a reference of shape (9165, 16624) using the following wrapped method:
+- platform normalization: platform_iterations=0, gene_keys=Cell type annotation, normalize_to=adata
   +- multi center: multi_center=10 multi_center_amplitudes=True
      +- bisection boost: bisections=4, bisection_divisor=3
         +- core: method=OT annotation_prior=None
mean,std( rescaling(gene) )  0.22523548816215352 0.4240753318419876
bisection run on 1
bisection run on 0.6666666666666667
bisection run on 0.4444444444444444
bisection run on 0.2962962962962963
bisection run on 0.19753086419753085
bisection run on 0.09876543209876543
Finished annotation in 14.39 seconds.
running method &#39;WOT&#39;
Starting preprocessing
Finished preprocessing in 2.03 seconds.
Starting annotation of data with shape (976, 16624) and a reference of shape (9165, 16624) using the following wrapped method:
+- multi center: multi_center=None multi_center_amplitudes=True
   +- core: method=WOT annotation_prior=None
Finished annotation in 28.16 seconds.
running method &#39;SVM&#39;
Starting preprocessing
Finished preprocessing in 2.06 seconds.
Starting annotation of data with shape (976, 16624) and a reference of shape (9165, 16624) using the following wrapped method:
+- multi center: multi_center=None multi_center_amplitudes=True
   +- core: method=svm annotation_prior=None
Finished annotation in 8.68 seconds.
running method &#39;NMFreg&#39;
Starting preprocessing
Finished preprocessing in 2.08 seconds.
Starting annotation of data with shape (976, 16624) and a reference of shape (9165, 16624) using the following wrapped method:
+- multi center: multi_center=None multi_center_amplitudes=True
   +- core: method=NMFreg annotation_prior=None
Finished annotation in 122.39 seconds.
running method &#39;SingleR&#39;
Starting preprocessing
Finished preprocessing in 2.05 seconds.
Starting annotation of data with shape (976, 16624) and a reference of shape (9165, 16624) using the following wrapped method:
+- multi center: multi_center=None multi_center_amplitudes=True
   +- core: method=SingleR annotation_prior=None conda_env=/ahg/regevdata/projects/mouse_CRC/rerun/clean_tacco_examples/results/env_links/SingleR_env
SCprep...time 0.2952430248260498
SCprep...time 1.15244722366333
[1] &#34;reference.h5ad&#34;       &#34;data.h5ad&#34;            &#34;result&#34;
[4] &#34;Cell type annotation&#34; &#34;True&#34;                 &#34;False&#34;
[7] &#34;de&#34;                   &#34;classic&#34;
[1] &#34;reading data&#34;
[1] &#34;csr&#34;
[1] &#34;reading reference&#34;
[1] &#34;csr&#34;
[1] &#34;running SingleR&#34;

Warning message:
package &#39;hdf5r&#39; was built under R version 4.1.1

Attaching package: &#39;SparseM&#39;

The following object is masked from &#39;package:base&#39;:

    backsolve

Warning message:
package &#39;SparseM&#39; was built under R version 4.1.3
Loading required package: SummarizedExperiment
Loading required package: MatrixGenerics
Loading required package: matrixStats

Attaching package: &#39;MatrixGenerics&#39;

The following objects are masked from &#39;package:matrixStats&#39;:

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars

Loading required package: GenomicRanges
Loading required package: stats4
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: &#39;BiocGenerics&#39;

The following objects are masked from &#39;package:parallel&#39;:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from &#39;package:stats&#39;:

    IQR, mad, sd, var, xtabs

The following objects are masked from &#39;package:base&#39;:

    Filter, Find, Map, Position, Reduce, anyDuplicated, append,
    as.data.frame, basename, cbind, colnames, dirname, do.call,
    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,
    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,
    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,
    tapply, union, unique, unsplit, which.max, which.min

Loading required package: S4Vectors

Attaching package: &#39;S4Vectors&#39;

The following objects are masked from &#39;package:Matrix&#39;:

    expand, unname

The following object is masked from &#39;package:hdf5r&#39;:

    values

The following objects are masked from &#39;package:data.table&#39;:

    first, second

The following objects are masked from &#39;package:base&#39;:

    I, expand.grid, unname

Loading required package: IRanges

Attaching package: &#39;IRanges&#39;

The following object is masked from &#39;package:data.table&#39;:

    shift

Loading required package: GenomeInfoDb
Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    &#39;browseVignettes()&#39;. To cite Bioconductor, see
    &#39;citation(&#34;Biobase&#34;)&#39;, and for packages &#39;citation(&#34;pkgname&#34;)&#39;.


Attaching package: &#39;Biobase&#39;

The following object is masked from &#39;package:MatrixGenerics&#39;:

    rowMedians

The following objects are masked from &#39;package:matrixStats&#39;:

    anyMissing, rowMedians

Warning message:
package &#39;matrixStats&#39; was built under R version 4.1.3

Finished annotation in 215.69 seconds.
running method &#39;RCTD&#39;
Starting preprocessing
Finished preprocessing in 2.0 seconds.
Starting annotation of data with shape (976, 16624) and a reference of shape (9165, 16624) using the following wrapped method:
+- multi center: multi_center=None multi_center_amplitudes=True
   +- core: method=RCTD annotation_prior=None conda_env=/ahg/regevdata/projects/mouse_CRC/rerun/clean_tacco_examples/results/env_links/RCTD_env
 [1] &#34;reference.h5ad&#34;       &#34;data.h5ad&#34;            &#34;result&#34;
 [4] &#34;0&#34;                    &#34;1&#34;                    &#34;full&#34;
 [7] &#34;Cell type annotation&#34; &#34;x&#34;                    &#34;y&#34;
[10] &#34;300&#34;
[1] &#34;reading data&#34;
[1] &#34;csr&#34;
[1] &#34;reading reference&#34;
[1] &#34;csr&#34;
[1] &#34;running RCTD&#34;

      Baso    Ccr7_DC        Eos  Erythroid   Lymphoid       Mast        Meg
      1430         24         59         87         98        318        272
  Monocyte Neutrophil        pDC
      3223       3642         12


Attaching package: &#39;SparseM&#39;

The following object is masked from &#39;package:base&#39;:

    backsolve

Begin: process_cell_type_info
process_cell_type_info: number of cells in reference: 9165
process_cell_type_info: number of genes in reference: 16624
End: process_cell_type_info
create.RCTD: getting regression differentially expressed genes:
get_de_genes: Baso found DE genes: 153
get_de_genes: Ccr7_DC found DE genes: 290
get_de_genes: Eos found DE genes: 112
get_de_genes: Erythroid found DE genes: 176
get_de_genes: Lymphoid found DE genes: 130
get_de_genes: Mast found DE genes: 97
get_de_genes: Meg found DE genes: 171
get_de_genes: Monocyte found DE genes: 180
get_de_genes: Neutrophil found DE genes: 139
get_de_genes: pDC found DE genes: 380
get_de_genes: total DE genes: 1475
create.RCTD: getting platform effect normalization differentially expressed genes:
get_de_genes: Baso found DE genes: 315
get_de_genes: Ccr7_DC found DE genes: 607
get_de_genes: Eos found DE genes: 281
get_de_genes: Erythroid found DE genes: 572
get_de_genes: Lymphoid found DE genes: 435
get_de_genes: Mast found DE genes: 292
get_de_genes: Meg found DE genes: 384
get_de_genes: Monocyte found DE genes: 363
get_de_genes: Neutrophil found DE genes: 304
get_de_genes: pDC found DE genes: 853
get_de_genes: total DE genes: 3109
fitBulk: decomposing bulk
chooseSigma: using initial Q_mat with sigma =  1
Likelihood value: 399948.20519735
Sigma value:  0.84
Likelihood value: 395804.961806008
Sigma value:  0.69
Likelihood value: 393229.045037818
Sigma value:  0.61
Likelihood value: 392458.511099675
Sigma value:  0.55
Likelihood value: 392177.634519394
Sigma value:  0.53
Likelihood value: 392141.960359694
Sigma value:  0.52
Likelihood value: 392135.082178793
Sigma value:  0.52

Finished annotation in 415.6 seconds.
</pre></div></div>
</div>
</section>
<section id="Plot-results">
<h2>Plot results<a class="headerlink" href="#Plot-results" title="Link to this heading">¶</a></h2>
<p>Plot gray background of day 2 data when showing annotation of day 4 and day 6 data and vice versa.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">joint</span> <span class="o">=</span> <span class="n">ad</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">d2</span><span class="p">,</span><span class="n">d4d6</span><span class="p">],)</span>
<span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">method_configs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
    <span class="n">joint</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">method</span><span class="p">]</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="n">method</span><span class="p">]</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">joint</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">joint</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;type_2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d2</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;clone_fate&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">joint</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">joint</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;type_46&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">d4d6</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;Cell type annotation&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">reindex</span><span class="p">(</span><span class="n">joint</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">dummy_key</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">utils</span><span class="o">.</span><span class="n">find_unused_key</span><span class="p">(</span><span class="n">joint</span><span class="o">.</span><span class="n">obs</span><span class="p">)</span>
<span class="n">joint</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="n">dummy_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">dummy_val</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">axs</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">)</span>

<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">dummy_key</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">only_gray</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;type_46&#39;</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>

<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">dummy_key</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">only_gray</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;type_2&#39;</span><span class="p">,</span>  <span class="n">colors</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>

<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">dummy_key</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">only_gray</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>  <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;TACCO&#39;</span><span class="p">,</span>  <span class="n">colors</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">);</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_single_cell_differentiation_15_0.png" src="../_images/notebooks_single_cell_differentiation_15_0.png" />
</div>
</div>
<p>Compare the correlations of all methods</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">corrs</span> <span class="o">=</span> <span class="p">{</span> <span class="n">method</span><span class="p">:</span> <span class="n">tc</span><span class="o">.</span><span class="n">ev</span><span class="o">.</span><span class="n">compute_err</span><span class="p">(</span><span class="n">d2</span><span class="p">,</span> <span class="n">method</span><span class="p">,</span> <span class="s1">&#39;clone_fate&#39;</span><span class="p">,</span> <span class="n">err_method</span><span class="o">=</span><span class="s1">&#39;corr&#39;</span><span class="p">)[</span><span class="n">method</span><span class="p">]</span> <span class="k">for</span> <span class="n">method</span> <span class="ow">in</span> <span class="n">method_configs</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="p">}</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span><span class="n">axs</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">axsize</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">])</span><span class="o">*</span><span class="mf">0.4</span><span class="p">)</span>
<span class="n">ticks</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">method_configs</span><span class="p">))</span>
<span class="n">colors</span> <span class="o">=</span> <span class="p">[</span><span class="n">common_code</span><span class="o">.</span><span class="n">method_color</span><span class="p">(</span><span class="n">m</span><span class="p">)</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">method_configs</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
<span class="n">ax</span> <span class="o">=</span> <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span>
<span class="n">ax</span><span class="o">.</span><span class="n">bar</span><span class="p">(</span><span class="n">ticks</span><span class="p">,</span> <span class="n">corrs</span><span class="o">.</span><span class="n">values</span><span class="p">(),</span> <span class="n">color</span><span class="o">=</span><span class="n">colors</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xticks</span><span class="p">([])</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_xmargin</span><span class="p">(</span><span class="mf">.1</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s1">&#39;Pearson correlation&#39;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">matplotlib.patches</span>
<span class="n">ax</span><span class="o">.</span><span class="n">legend</span><span class="p">(</span><span class="n">handles</span><span class="o">=</span><span class="p">[</span><span class="n">matplotlib</span><span class="o">.</span><span class="n">patches</span><span class="o">.</span><span class="n">Patch</span><span class="p">(</span><span class="n">color</span><span class="o">=</span><span class="n">color</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="n">ind</span><span class="p">)</span> <span class="k">for</span> <span class="p">(</span><span class="n">ind</span><span class="p">,</span> <span class="n">color</span><span class="p">)</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">method_configs</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="n">colors</span><span class="p">)</span> <span class="p">],</span>
    <span class="n">bbox_to_anchor</span><span class="o">=</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="n">loc</span><span class="o">=</span><span class="s1">&#39;upper left&#39;</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;matplotlib.legend.Legend at 0x2ae78bbcaa60&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_single_cell_differentiation_18_1.png" src="../_images/notebooks_single_cell_differentiation_18_1.png" />
</div>
</div>
<p>Show the annotations of all methods</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n_panels</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_configs</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>
<span class="n">fig</span><span class="p">,</span><span class="n">axs</span> <span class="o">=</span> <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">subplots</span><span class="p">((</span><span class="n">n_panels</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span><span class="o">//</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="n">axsize</span><span class="o">=</span><span class="n">axsize</span><span class="p">,</span> <span class="n">y_padding</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">axs</span> <span class="o">=</span> <span class="n">axs</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>

<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">dummy_key</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">only_gray</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="s1">&#39;type_2&#39;</span><span class="p">,</span>  <span class="n">colors</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="s1">&#39;clone differentiation&#39;</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i_method</span><span class="p">,</span><span class="n">method</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">method_configs</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">i_method</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">dummy_key</span><span class="p">,</span> <span class="n">colors</span><span class="o">=</span><span class="n">only_gray</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">tc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span><span class="n">joint</span><span class="p">,</span> <span class="n">joint</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">position_key</span><span class="o">=</span><span class="p">[</span><span class="n">x</span><span class="p">,</span><span class="n">y</span><span class="p">],</span> <span class="n">ax</span><span class="o">=</span><span class="n">axs</span><span class="p">[</span><span class="n">i_method</span><span class="o">+</span><span class="mi">1</span><span class="p">],</span> <span class="n">legend</span><span class="o">=</span><span class="p">(</span><span class="n">i_method</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_configs</span><span class="p">)</span><span class="o">-</span><span class="mi">1</span><span class="p">),</span>  <span class="n">point_size</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">keys</span><span class="o">=</span><span class="n">method</span><span class="p">,</span>  <span class="n">colors</span><span class="o">=</span><span class="n">color_dict</span><span class="p">,</span> <span class="n">noticks</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="n">i_method</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">set_title</span><span class="p">(</span><span class="n">method</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">method_configs</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">axs</span><span class="p">):</span>
    <span class="n">axs</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;off&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/notebooks_single_cell_differentiation_20_0.png" src="../_images/notebooks_single_cell_differentiation_20_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="simulated_dropout.html" class="btn btn-neutral float-left" title="Simulated single cell expression data with dropout" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="slideseq_mouse_colon.html" class="btn btn-neutral float-right" title="Slide-Seq Mouse Colon" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>