References
----------

.. [Wang21] Wang & Klein (2021),
   *Learning dynamics by computational integration of single cell genomic and lineage information*,
   `BioRxiv <https://www.biorxiv.org/content/10.1101/2021.05.06.443026v1>`__.

.. [Herman18] Herman, Sagar & Gr√ºn (2018),
   *FateID infers cell fate bias in multipotent progenitors from single-cell RNA-seq data*,
   `Nature Methods <https://doi.org/10.1038%2Fnmeth.4662>`__.

.. [Klopfenstein18] Klopfenstein *et al.* (2018),
   *GOATOOLS: A Python library for Gene Ontology analyses*,
   `Scientific Reports <https://www.nature.com/articles/s41598-018-28948-z>`__.

.. [Cable21] Cable *et al.* (2021),
   *Robust decomposition of cell type mixtures in spatial transcriptomics*,
   `Nature Biotechnology <https://www.nature.com/articles/s41587-021-00830-w>`__.
   
..
   .. [Bingham2018] Bingham *et al.* (2018),
      *Pyro: Deep Universal Probabilistic Programming*,
      `Journal of Machine Learning Research <https://www.jmlr.org/papers/volume20/18-403/18-403.pdf>`__.

.. [Rodriques19] Rodriques *et al.* (2019),
   *Slide-seq: A scalable technology for measuring genome-wide expression at high spatial resolution*,
   `Science <https://science.sciencemag.org/content/363/6434/1463>`__.

.. [Nitzan19] Nitzan *et al.* (2019),
   *Gene expression cartography*,
   `Nature <https://www.nature.com/articles/s41586-019-1773-3>`__.

.. [Aran19] Aran *et al.* (2019),
   *Reference-based analysis of lung single-cell sequencing reveals a transitional profibrotic macrophage*,
   `Nature Immunology <https://doi.org/10.1038/s41590-018-0276-y>`__.

.. [Abdelaal19] Abdelaal *et al.* (2019),
   *A comparison of automatic cell identification methods for single-cell RNA sequencing data*,
   `Genome Biology <https://doi.org/10.1186/s13059-019-1795-z>`__.

.. [Biancalani20] Biancalani *et al.* (2020),
   *Deep learning and alignment of spatially-resolved whole transcriptomes of single cells in the mouse brain with Tangram*,
   `BioRxiv <https://doi.org/10.1101/2020.08.29.272831>`__.

.. [Schiebinger19] Schiebinger *et al.* (2019),
   *Optimal-Transport Analysis of Single-Cell Gene Expression Identifies Developmental Trajectories in Reprogramming*,
   `Cell <10.1016/j.cell.2019.01.006>`__.

.. [Chizat16] Chizat *et al.* (2018).
   *Scaling algorithms for unbalanced optimal transport problems*,
   `Math. Comp. <https://doi.org/10.1090/mcom/3303>`__.

.. [Flamary2021] Flamary *et al.* (2021).
   *POT: Python Optimal Transport*,
   `Journal of Machine Learning Research <http://jmlr.org/papers/v22/20-451.html>`__.
